{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# <font color='blue'>Data Science Academy</font>\n",
    "# <font color='blue'>Matemática Para Machine Learning</font>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Lista de Exercícios  - Capítulo 5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "O objetivo desta lista de exercícios é você praticar os principais conceitos estudados neste capítulo, ao mesmo tempo que desenvolve suas habilidades em lógica de programação com a linguagem Python. \n",
    "\n",
    "Caso tenha dúvidas, isso é absolutamente normal e faça um trabalho de pesquisa a fim de relembrar o formato das operações matemáticas.\n",
    "\n",
    "Quando encontrar o formato de uma operação que resolva o exercício proposto, use a linguagem Python para representar esta operação. Em essência, é assim que aplicamos Matemática Para Machine Learning, construindo algoritmos e representando esses algoritmos em linguagem de programação."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A maioria das pessoas considera a Matemática e a Física, assustadoras bestas das quais é melhor manter a distância. Computadores, no entanto, podem nos ajudar a domar a complexidade e a aritmética tediosa das manipulações associadas a esses assuntos. De fato, Matemática e Física são muito mais acessíveis, uma vez que você tem o poder de computadores do seu lado.\n",
    "\n",
    "Divirta-se!!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercício 1 - Crie um algoritmo em Python para a multiplicação das duas matrizes abaixo. Seu algoritmo deve checar se as matrizes atendem às condições de multiplicação de matrizes estudadas neste capítulo.\n",
    "\n",
    "ATENÇÃO: NÃO use o NumPy (seria fácil demais). O objetivo é você testar sua lógica de programação para operação com matrizes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Solução\n",
    "\n",
    "# Matrizes\n",
    "x = [[1,2,3],[4,5,6],[7,8,9],[10,11,12]]\n",
    "y = [[1,2],[1,2],[3,4]]\n",
    "\n",
    "def matrixmult (A, B):\n",
    "    rows_A = len(A)\n",
    "    cols_A = len(A[0])\n",
    "    rows_B = len(B)\n",
    "    cols_B = len(B[0])\n",
    "\n",
    "    if cols_A != rows_B:\n",
    "        print(\"Não é possível multiplicar essas duas matrizes. Dimensões incorretas.\")\n",
    "        return\n",
    "\n",
    "    # Crie a matriz de resultados como uma matriz nula\n",
    "    # Dimensões devem ser rows_A x cols_B\n",
    "    C = [[0 for row in range(cols_B)] for col in range(rows_A)]\n",
    "    print(\"\\nMatriz nula criada para receber o resultado da operação: \", C)\n",
    "\n",
    "    for i in range(rows_A):\n",
    "        for j in range(cols_B):\n",
    "            for k in range(cols_A):\n",
    "                C[i][j] += A[i][k] * B[k][j]\n",
    "    return C\n",
    "\n",
    "matriz_resultante = matrixmult(x, y)\n",
    "print(\"\\nMatriz resultante da multiplicação entre as duas matrizes: \", matriz_resultante)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercício 2 - Agora sim, usando NumPy, verifique se o algoritmo criado no exercício anterior realmente está correto validando o resultado com o que o NumPy oferece para multiplicação de matrizes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "matrix([[12, 18],\n",
       "        [27, 42],\n",
       "        [42, 66],\n",
       "        [57, 90]])"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Solução\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "# Matrizes\n",
    "x = [[1,2,3],[4,5,6],[7,8,9],[10,11,12]]\n",
    "y = [[1,2],[1,2],[3,4]]\n",
    "\n",
    "# Converte as listas Python para matrizes NumPy\n",
    "mx = np.matrix(x)\n",
    "my = np.matrix(y)  \n",
    "\n",
    "# Resultado\n",
    "mx * my"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercício 3 - Considerando as duas matrizes abaixo, crie um algoritmo para multiplicar as matrizes.\n",
    "\n",
    "Dica: Use loops aninhados para compreender a regra matemática e traduzir isso em linguagem de programação."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[114, 160, 60, 27]\n",
      "[74, 97, 73, 14]\n",
      "[119, 157, 112, 23]\n"
     ]
    }
   ],
   "source": [
    "# Solução\n",
    "\n",
    "# Matriz 3x3 \n",
    "X = [[12,7,3],\n",
    "    [4 ,5,6],\n",
    "    [7 ,8,9]]\n",
    "\n",
    "# Matriz 3x4\n",
    "Y = [[5,8,1,2],\n",
    "    [6,7,3,0],\n",
    "    [4,5,9,1]]\n",
    "\n",
    "# Matriz nula 3x4\n",
    "result = [[0,0,0,0],\n",
    "         [0,0,0,0],\n",
    "         [0,0,0,0]]\n",
    "\n",
    "# Iteração pelas linhas de x\n",
    "for i in range(len(X)):\n",
    "   \n",
    "    # Iteração pelas colunas de y\n",
    "    for j in range(len(Y[0])):\n",
    "       \n",
    "        # Iteração pelas linhas de y\n",
    "        for k in range(len(Y)):\n",
    "            result[i][j] += X[i][k] * Y[k][j]\n",
    "\n",
    "# Imprime o resultado\n",
    "for r in result:\n",
    "    print(r)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercício 4 - Realize a mesma operação anterior, mas desta vez usando list comprehension."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[114, 160, 60, 27]\n",
      "[74, 97, 73, 14]\n",
      "[119, 157, 112, 23]\n"
     ]
    }
   ],
   "source": [
    "# Solução\n",
    "\n",
    "# Matriz 3x3 \n",
    "X = [[12,7,3],\n",
    "    [4 ,5,6],\n",
    "    [7 ,8,9]]\n",
    "\n",
    "# Matriz 3x4\n",
    "Y = [[5,8,1,2],\n",
    "    [6,7,3,0],\n",
    "    [4,5,9,1]]\n",
    "\n",
    "# Resultado 3x4\n",
    "result = [[sum(a*b for a,b in zip(X_row,Y_col)) for Y_col in zip(*Y)] for X_row in X]\n",
    "\n",
    "# Imprime o resultado\n",
    "for r in result:\n",
    "    print(r)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercício 5 - Considere as duas matrizes abaixo, x e y, e as duas diferentes soluções apresentadas para a multiplicação entre elas. Para cada execução, nós usamos o operador mágico %timeit para medir o tempo de execução. Você saberia explicar porque o tempo de execução é muito diferente? \n",
    "\n",
    "Dica: Essa costuma ser uma pergunta em entrevista para posição de Cientista de Dados."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Tempo de execução na multiplicação customizada:\n",
      "9.32 µs ± 135 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)\n",
      "\n",
      "Tempo de execução na multiplicação via NumPy:\n",
      "5.38 µs ± 25.4 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)\n"
     ]
    }
   ],
   "source": [
    "# Solução\n",
    "\n",
    "import numpy as np\n",
    "\n",
    "x = [[1,2,3],[4,5,6],[7,8,9]]\n",
    "y = [[1,2,3],[1,2,3],[3,4,5]]\n",
    "\n",
    "def matrixmult (A, B):\n",
    "    rows_A = len(A)\n",
    "    cols_A = len(A[0])\n",
    "    rows_B = len(B)\n",
    "    cols_B = len(B[0])\n",
    "\n",
    "    if cols_A != rows_B:\n",
    "        print(\"Não é possível multiplicar essas duas matrizes. Dimensões incorretas.\")\n",
    "        return\n",
    "\n",
    "    # Crie a matriz de resultados como uma matriz nula\n",
    "    # Dimensões devem ser rows_A x cols_B\n",
    "    C = [[0 for row in range(cols_B)] for col in range(rows_A)]\n",
    "\n",
    "    for i in range(rows_A):\n",
    "        for j in range(cols_B):\n",
    "            for k in range(cols_A):\n",
    "                C[i][j] += A[i][k] * B[k][j]\n",
    "    return C\n",
    "\n",
    "# Tempo de execução na multiplicação customizada\n",
    "print(\"\\nTempo de execução na multiplicação customizada:\")\n",
    "%timeit matrixmult(x, y)\n",
    "\n",
    "\n",
    "# Tempo de execução na multiplicação via NumPy\n",
    "print(\"\\nTempo de execução na multiplicação via NumPy:\")\n",
    "%timeit np.multiply(x, y)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Resposta: As funções do NumPy são otimizadas para performance e não precisamos \"reinventar a roda\", tentando criar o que já existe!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercício 6 - Calcule o determinante da matriz abaixo, usando NumPy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "-2.0000000000000004\n"
     ]
    }
   ],
   "source": [
    "# Solução\n",
    "\n",
    "import numpy\n",
    "\n",
    "# Matriz\n",
    "M = [[1, 2], [3, 4]]\n",
    "\n",
    "# Determinante\n",
    "print (numpy.linalg.det(M))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercício 7 - Calcule a inversa da matriz abaixo usando NumPy."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Solução 1:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "matrix([[-2.5,  1.5],\n",
       "        [ 2. , -1. ]])"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Solução\n",
    "\n",
    "x = [[2,3],[4,5]]\n",
    "\n",
    "m = np.matrix(x)\n",
    "\n",
    "# Solução 1\n",
    "print(\"\\nSolução 1:\")\n",
    "m.I"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Solução 2:\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "matrix([[-2.5,  1.5],\n",
       "        [ 2. , -1. ]])"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Solução 2\n",
    "print(\"\\nSolução 2:\")\n",
    "np.linalg.inv(m)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercício 8 - O TensorFlow é o principal framework para Deep Learning. Usando uma de suas funções, encontre a inversa da matriz abaixo.\n",
    "\n",
    "Referência: https://www.tensorflow.org/api_guides/python/math_ops#Matrix_Math_Functions\n",
    "\n",
    "Dica: Pode ser necessário instalar o TensorFlow com o comando abaixo (descomente a linha e execute a célula)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: tensorflow in /anaconda3/lib/python3.6/site-packages (1.10.1)\n",
      "Requirement already satisfied: protobuf>=3.6.0 in /anaconda3/lib/python3.6/site-packages (from tensorflow) (3.6.1)\n",
      "Requirement already satisfied: astor>=0.6.0 in /anaconda3/lib/python3.6/site-packages (from tensorflow) (0.7.1)\n",
      "Requirement already satisfied: gast>=0.2.0 in /anaconda3/lib/python3.6/site-packages (from tensorflow) (0.2.0)\n",
      "Requirement already satisfied: grpcio>=1.8.6 in /anaconda3/lib/python3.6/site-packages (from tensorflow) (1.15.0)\n",
      "Requirement already satisfied: six>=1.10.0 in /anaconda3/lib/python3.6/site-packages (from tensorflow) (1.11.0)\n",
      "Requirement already satisfied: numpy<=1.14.5,>=1.13.3 in /anaconda3/lib/python3.6/site-packages (from tensorflow) (1.14.3)\n",
      "Requirement already satisfied: setuptools<=39.1.0 in /anaconda3/lib/python3.6/site-packages (from tensorflow) (39.1.0)\n",
      "Requirement already satisfied: tensorboard<1.11.0,>=1.10.0 in /anaconda3/lib/python3.6/site-packages (from tensorflow) (1.10.0)\n",
      "Requirement already satisfied: wheel>=0.26 in /anaconda3/lib/python3.6/site-packages (from tensorflow) (0.31.1)\n",
      "Requirement already satisfied: absl-py>=0.1.6 in /anaconda3/lib/python3.6/site-packages (from tensorflow) (0.4.1)\n",
      "Requirement already satisfied: termcolor>=1.1.0 in /anaconda3/lib/python3.6/site-packages (from tensorflow) (1.1.0)\n",
      "Requirement already satisfied: markdown>=2.6.8 in /anaconda3/lib/python3.6/site-packages (from tensorboard<1.11.0,>=1.10.0->tensorflow) (2.6.11)\n",
      "Requirement already satisfied: werkzeug>=0.11.10 in /anaconda3/lib/python3.6/site-packages (from tensorboard<1.11.0,>=1.10.0->tensorflow) (0.14.1)\n",
      "\u001b[31mdistributed 1.21.8 requires msgpack, which is not installed.\u001b[0m\n",
      "\u001b[33mYou are using pip version 10.0.1, however version 18.0 is available.\n",
      "You should consider upgrading via the 'pip install --upgrade pip' command.\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "!pip install tensorflow"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n",
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 18, in swig_import_helper\n",
      "    fp, pathname, description = imp.find_module('_pywrap_tensorflow_internal', [dirname(__file__)])\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/imp.py\", line 296, in find_module\n",
      "    raise ImportError(_ERR_MSG.format(name), name=name)\n",
      "ImportError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\", line 58, in <module>\n",
      "    from tensorflow.python.pywrap_tensorflow_internal import *\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 28, in <module>\n",
      "    _pywrap_tensorflow_internal = swig_import_helper()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 20, in swig_import_helper\n",
      "    import _pywrap_tensorflow_internal\n",
      "ModuleNotFoundError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3326, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-1-f5c174befec4>\", line 3, in <module>\n",
      "    import tensorflow as tf\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 98, in <module>\n",
      "    from tensorflow_core import *\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/__init__.py\", line 40, in <module>\n",
      "    from tensorflow.python.tools import module_util as _module_util\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/importlib/__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/__init__.py\", line 49, in <module>\n",
      "    from tensorflow.python import pywrap_tensorflow\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\", line 74, in <module>\n",
      "    raise ImportError(msg)\n",
      "ImportError: Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 18, in swig_import_helper\n",
      "    fp, pathname, description = imp.find_module('_pywrap_tensorflow_internal', [dirname(__file__)])\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/imp.py\", line 296, in find_module\n",
      "    raise ImportError(_ERR_MSG.format(name), name=name)\n",
      "ImportError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\", line 58, in <module>\n",
      "    from tensorflow.python.pywrap_tensorflow_internal import *\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 28, in <module>\n",
      "    _pywrap_tensorflow_internal = swig_import_helper()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 20, in swig_import_helper\n",
      "    import _pywrap_tensorflow_internal\n",
      "ModuleNotFoundError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "\n",
      "Failed to load the native TensorFlow runtime.\n",
      "\n",
      "See https://www.tensorflow.org/install/errors\n",
      "\n",
      "for some common reasons and solutions.  Include the entire stack trace\n",
      "above this error message when asking for help.\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2040, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'ImportError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 18, in swig_import_helper\n",
      "    fp, pathname, description = imp.find_module('_pywrap_tensorflow_internal', [dirname(__file__)])\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/imp.py\", line 296, in find_module\n",
      "    raise ImportError(_ERR_MSG.format(name), name=name)\n",
      "ImportError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\", line 58, in <module>\n",
      "    from tensorflow.python.pywrap_tensorflow_internal import *\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 28, in <module>\n",
      "    _pywrap_tensorflow_internal = swig_import_helper()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 20, in swig_import_helper\n",
      "    import _pywrap_tensorflow_internal\n",
      "ModuleNotFoundError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 319, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 353, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/inspect.py\", line 1502, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/inspect.py\", line 1460, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/inspect.py\", line 733, in getmodule\n",
      "    if ismodule(module) and hasattr(module, '__file__'):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/importlib/__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 953, in _find_and_load_unlocked\n",
      "  File \"<frozen importlib._bootstrap>\", line 219, in _call_with_frames_removed\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 967, in _find_and_load_unlocked\n",
      "  File \"<frozen importlib._bootstrap>\", line 677, in _load_unlocked\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 728, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 219, in _call_with_frames_removed\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/__init__.py\", line 42, in <module>\n",
      "    from . _api.v2 import audio\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/_api/v2/audio/__init__.py\", line 10, in <module>\n",
      "    from tensorflow.python.ops.gen_audio_ops import decode_wav\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/ops/gen_audio_ops.py\", line 10, in <module>\n",
      "    from tensorflow.python import pywrap_tensorflow as _pywrap_tensorflow\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/importlib/__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/__init__.py\", line 49, in <module>\n",
      "    from tensorflow.python import pywrap_tensorflow\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\", line 74, in <module>\n",
      "    raise ImportError(msg)\n",
      "ImportError: Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 18, in swig_import_helper\n",
      "    fp, pathname, description = imp.find_module('_pywrap_tensorflow_internal', [dirname(__file__)])\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/imp.py\", line 296, in find_module\n",
      "    raise ImportError(_ERR_MSG.format(name), name=name)\n",
      "ImportError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\", line 58, in <module>\n",
      "    from tensorflow.python.pywrap_tensorflow_internal import *\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 28, in <module>\n",
      "    _pywrap_tensorflow_internal = swig_import_helper()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 20, in swig_import_helper\n",
      "    import _pywrap_tensorflow_internal\n",
      "ModuleNotFoundError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3326, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-1-f5c174befec4>\", line 3, in <module>\n",
      "    import tensorflow as tf\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 98, in <module>\n",
      "    from tensorflow_core import *\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/__init__.py\", line 40, in <module>\n",
      "    from tensorflow.python.tools import module_util as _module_util\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/importlib/__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/__init__.py\", line 49, in <module>\n",
      "    from tensorflow.python import pywrap_tensorflow\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\", line 74, in <module>\n",
      "    raise ImportError(msg)\n",
      "ImportError: Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 18, in swig_import_helper\n",
      "    fp, pathname, description = imp.find_module('_pywrap_tensorflow_internal', [dirname(__file__)])\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/imp.py\", line 296, in find_module\n",
      "    raise ImportError(_ERR_MSG.format(name), name=name)\n",
      "ImportError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\", line 58, in <module>\n",
      "    from tensorflow.python.pywrap_tensorflow_internal import *\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 28, in <module>\n",
      "    _pywrap_tensorflow_internal = swig_import_helper()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 20, in swig_import_helper\n",
      "    import _pywrap_tensorflow_internal\n",
      "ModuleNotFoundError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "\n",
      "Failed to load the native TensorFlow runtime.\n",
      "\n",
      "See https://www.tensorflow.org/install/errors\n",
      "\n",
      "for some common reasons and solutions.  Include the entire stack trace\n",
      "above this error message when asking for help.\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2040, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'ImportError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 18, in swig_import_helper\n",
      "    fp, pathname, description = imp.find_module('_pywrap_tensorflow_internal', [dirname(__file__)])\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/imp.py\", line 296, in find_module\n",
      "    raise ImportError(_ERR_MSG.format(name), name=name)\n",
      "ImportError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\", line 58, in <module>\n",
      "    from tensorflow.python.pywrap_tensorflow_internal import *\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 28, in <module>\n",
      "    _pywrap_tensorflow_internal = swig_import_helper()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 20, in swig_import_helper\n",
      "    import _pywrap_tensorflow_internal\n",
      "ModuleNotFoundError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "\n",
      "Failed to load the native TensorFlow runtime.\n",
      "\n",
      "See https://www.tensorflow.org/install/errors\n",
      "\n",
      "for some common reasons and solutions.  Include the entire stack trace\n",
      "above this error message when asking for help.\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 18, in swig_import_helper\n",
      "    fp, pathname, description = imp.find_module('_pywrap_tensorflow_internal', [dirname(__file__)])\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/imp.py\", line 296, in find_module\n",
      "    raise ImportError(_ERR_MSG.format(name), name=name)\n",
      "ImportError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\", line 58, in <module>\n",
      "    from tensorflow.python.pywrap_tensorflow_internal import *\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 28, in <module>\n",
      "    _pywrap_tensorflow_internal = swig_import_helper()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 20, in swig_import_helper\n",
      "    import _pywrap_tensorflow_internal\n",
      "ModuleNotFoundError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3326, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-1-f5c174befec4>\", line 3, in <module>\n",
      "    import tensorflow as tf\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 98, in <module>\n",
      "    from tensorflow_core import *\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/__init__.py\", line 40, in <module>\n",
      "    from tensorflow.python.tools import module_util as _module_util\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/importlib/__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/__init__.py\", line 49, in <module>\n",
      "    from tensorflow.python import pywrap_tensorflow\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\", line 74, in <module>\n",
      "    raise ImportError(msg)\n",
      "ImportError: Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 18, in swig_import_helper\n",
      "    fp, pathname, description = imp.find_module('_pywrap_tensorflow_internal', [dirname(__file__)])\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/imp.py\", line 296, in find_module\n",
      "    raise ImportError(_ERR_MSG.format(name), name=name)\n",
      "ImportError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\", line 58, in <module>\n",
      "    from tensorflow.python.pywrap_tensorflow_internal import *\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 28, in <module>\n",
      "    _pywrap_tensorflow_internal = swig_import_helper()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 20, in swig_import_helper\n",
      "    import _pywrap_tensorflow_internal\n",
      "ModuleNotFoundError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "\n",
      "Failed to load the native TensorFlow runtime.\n",
      "\n",
      "See https://www.tensorflow.org/install/errors\n",
      "\n",
      "for some common reasons and solutions.  Include the entire stack trace\n",
      "above this error message when asking for help.\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2040, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'ImportError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3249, in run_ast_nodes\n",
      "    if (await self.run_code(code, result,  async_=asy)):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3343, in run_code\n",
      "    self.showtraceback(running_compiled_code=True)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2043, in showtraceback\n",
      "    value, tb, tb_offset=tb_offset)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1385, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1288, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1150, in structured_traceback\n",
      "    formatted_exceptions += self.prepare_chained_exception_message(evalue.__cause__)\n",
      "TypeError: can only concatenate str (not \"list\") to str\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2040, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 18, in swig_import_helper\n",
      "    fp, pathname, description = imp.find_module('_pywrap_tensorflow_internal', [dirname(__file__)])\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/imp.py\", line 296, in find_module\n",
      "    raise ImportError(_ERR_MSG.format(name), name=name)\n",
      "ImportError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\", line 58, in <module>\n",
      "    from tensorflow.python.pywrap_tensorflow_internal import *\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 28, in <module>\n",
      "    _pywrap_tensorflow_internal = swig_import_helper()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 20, in swig_import_helper\n",
      "    import _pywrap_tensorflow_internal\n",
      "ModuleNotFoundError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 319, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 353, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/inspect.py\", line 1502, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/inspect.py\", line 1460, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/inspect.py\", line 733, in getmodule\n",
      "    if ismodule(module) and hasattr(module, '__file__'):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/importlib/__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 953, in _find_and_load_unlocked\n",
      "  File \"<frozen importlib._bootstrap>\", line 219, in _call_with_frames_removed\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 967, in _find_and_load_unlocked\n",
      "  File \"<frozen importlib._bootstrap>\", line 677, in _load_unlocked\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 728, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 219, in _call_with_frames_removed\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/__init__.py\", line 42, in <module>\n",
      "    from . _api.v2 import audio\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/_api/v2/audio/__init__.py\", line 10, in <module>\n",
      "    from tensorflow.python.ops.gen_audio_ops import decode_wav\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/ops/gen_audio_ops.py\", line 10, in <module>\n",
      "    from tensorflow.python import pywrap_tensorflow as _pywrap_tensorflow\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/importlib/__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/__init__.py\", line 49, in <module>\n",
      "    from tensorflow.python import pywrap_tensorflow\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\", line 74, in <module>\n",
      "    raise ImportError(msg)\n",
      "ImportError: Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 18, in swig_import_helper\n",
      "    fp, pathname, description = imp.find_module('_pywrap_tensorflow_internal', [dirname(__file__)])\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/imp.py\", line 296, in find_module\n",
      "    raise ImportError(_ERR_MSG.format(name), name=name)\n",
      "ImportError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\", line 58, in <module>\n",
      "    from tensorflow.python.pywrap_tensorflow_internal import *\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 28, in <module>\n",
      "    _pywrap_tensorflow_internal = swig_import_helper()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 20, in swig_import_helper\n",
      "    import _pywrap_tensorflow_internal\n",
      "ModuleNotFoundError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3326, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-1-f5c174befec4>\", line 3, in <module>\n",
      "    import tensorflow as tf\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 98, in <module>\n",
      "    from tensorflow_core import *\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/__init__.py\", line 40, in <module>\n",
      "    from tensorflow.python.tools import module_util as _module_util\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/importlib/__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/__init__.py\", line 49, in <module>\n",
      "    from tensorflow.python import pywrap_tensorflow\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\", line 74, in <module>\n",
      "    raise ImportError(msg)\n",
      "ImportError: Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 18, in swig_import_helper\n",
      "    fp, pathname, description = imp.find_module('_pywrap_tensorflow_internal', [dirname(__file__)])\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/imp.py\", line 296, in find_module\n",
      "    raise ImportError(_ERR_MSG.format(name), name=name)\n",
      "ImportError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\", line 58, in <module>\n",
      "    from tensorflow.python.pywrap_tensorflow_internal import *\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 28, in <module>\n",
      "    _pywrap_tensorflow_internal = swig_import_helper()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 20, in swig_import_helper\n",
      "    import _pywrap_tensorflow_internal\n",
      "ModuleNotFoundError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "\n",
      "Failed to load the native TensorFlow runtime.\n",
      "\n",
      "See https://www.tensorflow.org/install/errors\n",
      "\n",
      "for some common reasons and solutions.  Include the entire stack trace\n",
      "above this error message when asking for help.\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2040, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'ImportError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3249, in run_ast_nodes\n",
      "    if (await self.run_code(code, result,  async_=asy)):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3343, in run_code\n",
      "    self.showtraceback(running_compiled_code=True)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2043, in showtraceback\n",
      "    value, tb, tb_offset=tb_offset)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1385, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1288, in structured_traceback\n",
      "    self, etype, value, tb, tb_offset, number_of_lines_of_context\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1150, in structured_traceback\n",
      "    formatted_exceptions += self.prepare_chained_exception_message(evalue.__cause__)\n",
      "TypeError: can only concatenate str (not \"list\") to str\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2040, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'TypeError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 18, in swig_import_helper\n",
      "    fp, pathname, description = imp.find_module('_pywrap_tensorflow_internal', [dirname(__file__)])\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/imp.py\", line 296, in find_module\n",
      "    raise ImportError(_ERR_MSG.format(name), name=name)\n",
      "ImportError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\", line 58, in <module>\n",
      "    from tensorflow.python.pywrap_tensorflow_internal import *\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 28, in <module>\n",
      "    _pywrap_tensorflow_internal = swig_import_helper()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 20, in swig_import_helper\n",
      "    import _pywrap_tensorflow_internal\n",
      "ModuleNotFoundError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "\n",
      "Failed to load the native TensorFlow runtime.\n",
      "\n",
      "See https://www.tensorflow.org/install/errors\n",
      "\n",
      "for some common reasons and solutions.  Include the entire stack trace\n",
      "above this error message when asking for help.\n"
     ]
    },
    {
     "ename": "TypeError",
     "evalue": "can only concatenate str (not \"list\") to str",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mImportError\u001b[0m                               Traceback (most recent call last)",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\u001b[0m in \u001b[0;36mswig_import_helper\u001b[0;34m()\u001b[0m\n\u001b[1;32m     17\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 18\u001b[0;31m             \u001b[0mfp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mpathname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdescription\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mimp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfind_module\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'_pywrap_tensorflow_internal'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mdirname\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m__file__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     19\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mImportError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/imp.py\u001b[0m in \u001b[0;36mfind_module\u001b[0;34m(name, path)\u001b[0m\n\u001b[1;32m    295\u001b[0m     \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 296\u001b[0;31m         \u001b[0;32mraise\u001b[0m \u001b[0mImportError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_ERR_MSG\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    297\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mImportError\u001b[0m: No module named '_pywrap_tensorflow_internal'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     57\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 58\u001b[0;31m   \u001b[0;32mfrom\u001b[0m \u001b[0mtensorflow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpython\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpywrap_tensorflow_internal\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     59\u001b[0m   \u001b[0;32mfrom\u001b[0m \u001b[0mtensorflow\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpython\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpywrap_tensorflow_internal\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0m__version__\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     27\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0m_mod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 28\u001b[0;31m     \u001b[0m_pywrap_tensorflow_internal\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mswig_import_helper\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     29\u001b[0m     \u001b[0;32mdel\u001b[0m \u001b[0mswig_import_helper\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\u001b[0m in \u001b[0;36mswig_import_helper\u001b[0;34m()\u001b[0m\n\u001b[1;32m     19\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mImportError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 20\u001b[0;31m             \u001b[0;32mimport\u001b[0m \u001b[0m_pywrap_tensorflow_internal\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     21\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0m_pywrap_tensorflow_internal\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named '_pywrap_tensorflow_internal'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\u001b[0m in \u001b[0;36mshowtraceback\u001b[0;34m(self, exc_tuple, filename, tb_offset, exception_only, running_compiled_code)\u001b[0m\n\u001b[1;32m   2039\u001b[0m                         \u001b[0;31m# in the engines. This should return a list of strings.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2040\u001b[0;31m                         \u001b[0mstb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_render_traceback_\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2041\u001b[0m                     \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'ImportError' object has no attribute '_render_traceback_'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mTypeError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\u001b[0m in \u001b[0;36mrun_code\u001b[0;34m(self, code_obj, result, async_)\u001b[0m\n\u001b[1;32m   3341\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mresult\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3342\u001b[0m                 \u001b[0mresult\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0merror_in_exec\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mexc_info\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3343\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshowtraceback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrunning_compiled_code\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3344\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3345\u001b[0m             \u001b[0moutflag\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\u001b[0m in \u001b[0;36mshowtraceback\u001b[0;34m(self, exc_tuple, filename, tb_offset, exception_only, running_compiled_code)\u001b[0m\n\u001b[1;32m   2041\u001b[0m                     \u001b[0;32mexcept\u001b[0m \u001b[0mException\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2042\u001b[0m                         stb = self.InteractiveTB.structured_traceback(etype,\n\u001b[0;32m-> 2043\u001b[0;31m                                             value, tb, tb_offset=tb_offset)\n\u001b[0m\u001b[1;32m   2044\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2045\u001b[0m                     \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_showtraceback\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0metype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstb\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/IPython/core/ultratb.py\u001b[0m in \u001b[0;36mstructured_traceback\u001b[0;34m(self, etype, value, tb, tb_offset, number_of_lines_of_context)\u001b[0m\n\u001b[1;32m   1383\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtb\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1384\u001b[0m         return FormattedTB.structured_traceback(\n\u001b[0;32m-> 1385\u001b[0;31m             self, etype, value, tb, tb_offset, number_of_lines_of_context)\n\u001b[0m\u001b[1;32m   1386\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1387\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/IPython/core/ultratb.py\u001b[0m in \u001b[0;36mstructured_traceback\u001b[0;34m(self, etype, value, tb, tb_offset, number_of_lines_of_context)\u001b[0m\n\u001b[1;32m   1286\u001b[0m             \u001b[0;31m# Verbose modes need a full traceback\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1287\u001b[0m             return VerboseTB.structured_traceback(\n\u001b[0;32m-> 1288\u001b[0;31m                 \u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0metype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtb\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtb_offset\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnumber_of_lines_of_context\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1289\u001b[0m             )\n\u001b[1;32m   1290\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mmode\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'Minimal'\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/opt/anaconda3/lib/python3.7/site-packages/IPython/core/ultratb.py\u001b[0m in \u001b[0;36mstructured_traceback\u001b[0;34m(self, etype, evalue, etb, tb_offset, number_of_lines_of_context)\u001b[0m\n\u001b[1;32m   1148\u001b[0m         \u001b[0mexception\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_parts_of_chained_exception\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mevalue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1149\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mexception\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1150\u001b[0;31m             \u001b[0mformatted_exceptions\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mprepare_chained_exception_message\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mevalue\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__cause__\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1151\u001b[0m             \u001b[0metype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mevalue\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0metb\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mexception\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1152\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mTypeError\u001b[0m: can only concatenate str (not \"list\") to str"
     ]
    }
   ],
   "source": [
    "# Solução\n",
    "\n",
    "import tensorflow as tf\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "x = [[2,3],[4,5]]\n",
    "\n",
    "x_cast = tf.cast(x, tf.float64)\n",
    "x_final = tf.linalg.inv(x_cast)\n",
    "print(x_final)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercício 9 (Desafio) - Crie um algoritmo de Machine Learning de Regressão Linear Simples, usando NumPy, que mostre a relação entre duas variáveis x e y (que são criadas abaixo usando valores aleatórios). Seu script deve conter as operações matemáticas de multiplicação de matrizes estudadas neste capítulo.\n",
    "\n",
    "Dica: Deixamos algumas linhas de código para ajudá-lo nesta tarefa."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(200,)\n",
      "(200,)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.legend.Legend at 0xb2b5175f8>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAbsAAAD8CAYAAADuQwp+AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAIABJREFUeJzt3XuUVdWdJ/Dv796CehiNJhZavBEBLYwQqRjsPHQi7VLDmClshOgoZkx4TfdK1gyrO8a1OlkhWT0ZXT090y0gY8dH2onEhKCdMavBGBOzVjAWDigPUVKAAUoojSG6gMKifvPHvjd1udzHeexzzj7nfj9r3VVV1K1z993ncn5n7/3be4uqgoiIKMtySReAiIgoagx2RESUeQx2RESUeQx2RESUeQx2RESUeQx2RESUeQx2RESUedaCnYjkReT/ichPbB2TiIjIBpstuy8D2GXxeERERFY02TiIiIwF8FkA3wbwX+o9//zzz9eJEyfaeGkiooaxZcuWt1S1PelypJGVYAfgHwD8NYCzvTx54sSJ6OnpsfTSRESNQUT2J12GtArdjSkicwEcUdUtdZ63WER6RKSnv78/7MsSERF5ZmPM7hMAbhKRfQAeB/AZEfmX8iep6lpV7VLVrvZ2tsKJiCg+oYOdqt6tqmNVdSKAhQCeVdX/GLpkREREltgasyMiogRs2bJlVFNT04MALkPjzp0eArB9cHDwi7NmzTpS6QlWg52qPgfgOZvHJCKi6pqamh688MILL21vb38nl8s15AalQ0ND0t/f3/nmm28+COCmSs9p1LsAInJRXx9w9dXAm28mXZI0uay9vf2PjRroACCXy2l7e/tRmNZt5efEWB4isilNgcFrWVeuBH71K+Cb34ynXNmQa+RAV1Sog6oxjcGOKK3SFBjqlbW1FRABVq8GhobMVxHz72mQphuPBsVgR5Q2aQoMXsva2wvceivQ1mZ+bmsDbrsN2Ls3/jIHUS+YZzwY5vP5WZdccknnxRdfPH3atGmd3/jGNy44deqUr2PcfPPNEx966KHzIioigx1R6qQpMHgta0cHcM45wIkTQEuL+XrOOcCFF8ZfZj+8BnPXWuH794/Axz42DW+8YSVJsbm5eejVV1/duWfPnh3PPvvsaxs3bvzgihUrRts4ti0MdkRpk6bA4Keshw8DS5cCmzebr2loBdUL5q62wu+5pwNbtnwAX/ua9YA0ZsyYwQcffHDfQw89NGpoaAi7d+8eOWvWrGmdnZ2XdnZ2Xrpp06azAGBoaAh33HHH+MmTJ0+/5pprLn7rrbf+FHiffPLJsy+99NLOqVOnds6fP3/i8ePHBQCWL18+ZvLkydOnTp3auXjx4rF+ysV5dkRpVAwMixcDa9eabjJXeS3r+vXD399/fzxlC6teMO/tBVasADZsAI4dM8Gwuxu4775kytvcfAVOnpQ//fzYY+147LF2jBypGBh4ydbLdHZ2nhwaGsLBgwebRo8ePfj888+/1tbWpq+88krz5z//+Yu2b9++63vf+965e/bsad69e/eOAwcOjPjIRz4y/c4773z72LFjsmTJkkkbN27cffnllw90d3dPvPfee9uXLFny9tNPP31eb2/v9lwuh7feeivvp0wMdkRplKbAkKayBlErmLvWCn/ttVfwV381Fps2nYsTJ3JoaRnCddf9Af/4j7+z/VKqJkH05MmTctddd03YuXNnay6Xw/79+5sB4Be/+MXZt9xyy++bmpowceLE96+66qp3AWDbtm0tY8eOHbj88ssHAODOO+98+/777x919913H2lubh5auHDhhM9+9rNHFyxYcNRPeRjsiIjCqBfMXWqFT5jwPs4++xQGBnKF1lwOZ599CuPHD9p8mZ07d47M5/MYM2bM4IoVK0aPGjXq/R/96Ed7h4aG0NraOqv4PBE542+LQbLciBEjsHXr1l1PPfXUOY8//vh5q1evHrV58+bXvJaJY3ZEZGQ8YzAx69ebIDhjhvlaGhyT0N8/Arfd1o/nntuF227rx5EjI2we/tChQ01f+tKXJnzhC184ksvlcPTo0XxHR8f7+Xweq1at+nAxS/Pqq69+94knnvjQ4OAg9u/fP2Lz5s1nA8DMmTNPHDx4cOT27dubAeDRRx/98Kc+9al3jx49mvv973+fX7BgwdE1a9b8bteuXW1+ysWWHREZpRmDq1YlXRqKysaNv/3T91dd9YaNQw4MDOQuueSSzsHBQcnn87pgwYK3v/71rx8GgK985StHbr755skbNmw475Of/OS7ra2tQwBw++23/+FnP/vZOdOmTZs+adKkE1deeeW7ANDW1qZr1qzZN3/+/MmnTp3CjBkzjq1YsaL/yJEjTXPnzr14YGBAAOBb3/qWr65XqdZkjFJXV5dy81YiR7S2mrGkci0twPHj8ZeHqhKRLaraVfpv27Zt2zdjxoy3kiqTS7Zt23b+jBkzJlb6HbsxiRpdmubtEQXEYEfU6FzLGCSKAIMdEaVzQjcVDQ0NDZ2Z1thgCnUwVO33TFAhouzPhYtCXx+wcCGwbl3SreDt/f39ne3t7UcbdfeDwn52HwSwvdpzQgc7EWkB8EsAzYXj/VBVvx72uERETnMke3VwcPCLb7755oNvvvkmdyofHPxitSeEzsYUMyvwLFV9T0RGAPgVgC+r6uZqf8NsTCJKrQSzVytlY5I3oe8C1Hiv8OOIwqMhm9JE1ACYvZpKVpq8IpIXka0AjgDYpKovVHjOYhHpEZGe/v5+Gy9LRBQ/Zq+mkpVgp6qnVHUmgLEArhSRyyo8Z62qdqlqV3t7u42XJSJKRmn26u23Az/4ATNYHWd1MFNV/wDgOQDX2zwuESWM62aernS9y7Y24J133NmYlSoKHexEpF1Ezi183wpgDoBXwx6XiBzi2k7btgUJ5q5uzEoV2WjZdQD4uYi8DOBFmDG7n1g4LhElrVEu6EGCORNVUsVGNubLqvpRVb1cVS9T1Yze+hE1oKxf0MMEcyaqpEqjTkAkIi+yfkEPG8y5zFpqcLkwIqrNpZ22bQsbzLnMWmow2BFRbVm/oGc5mNOfMNgRUWOrFMzdWeSZLOGYHRFRuaxPtWhADHZElA5xTGxvlKkWDYjBjojSIY7WVtanWjQwBjsiikfQllmcra2sT7VoYAx2RBSPoC2zuFtbnDuXSQx2RI0ozoWdw7bM4m5tlS7yfP/9p2dresFFs53EYEeUZkEvrHFmG9pomSXV2gpSv8zkdJKoxr+peFdXl/b09MT+ukSZs3w58MADwJIlwKpV9Z/f2mpaRuVaWoDjx+2Xr2jZMjNhe+RI4ORJ7+VN2vLlwJo1phX50ku1W5Mx1K2IbFHVLisHazBs2RGlUdCuwaSyDdM2DlZav6qmhdfRUbt+mcnpNK6gQpRGvb3AihXAhg3AsWPmwtrdDdx3X+2/SyrbMG1LjvX2AmPHmhuJUidOmIBXqaXGTE6nsWVHlEZ+L6ylY09pa2XZ4mf8raPDtMpKNTXVb6mV1u3ttwM/+EHj1K/j2LIjSis/CxiXJk2krZVlS2kdeBkvfO89YPp0YOdOIJcDBgfrt9RK67atDXjnHe+vR5EKnaAiIuMAPArgQgBDANaq6v+s9TdMUCGKSVIJKS4JUwfz5plWXukNxfr1tReKjrDOmaASnI1uzEEA/1VVLwUwG8B/FpFOC8clorCYNBGuDqrNuas1vYB17qTQwU5V+1T1pcL37wLYBWBM2OMSkQWNljRRaVzOVh309QH5fP0s2Ear85SwmqAiIhMBfBTACxV+t1hEekSkp7+/3+bLElE1fX3AE0+YZIk0JaTYnixvIyln5UozDWHKlPqttkZNAnKYtUnlIvIBAL8A8G1Vrbm+DsfsiGLid9K5K1yaLF/t2IBJXImxbjlmF5yVlp2IjADwIwCP1Qt0RBSDtO7LFqTcfX1mTK27O5pxsvIxuFzOtO6eeYatthQJHexERAD8M4Bdqvr34YtERKG5niRRrZsySLlXrgRefBHYvTuacbLyMTgAmDMHuPbaYAtFUyJstOw+AeB2AJ8Rka2Fx40WjktEQbmeJFFtbM1PuctbgTt3mq9DQ/ZbXByDSz0uBE3pU2uOEw2rNkcsSV7G1ryWu6+v+pJpGf1ccMwuOC4XRunjyhYqru9bFnZftih46ab0Wu6oWq+un1cKhMGO0sO1pAtXgm6a2A5QUXQv8rxmErsxKT1c6bbiElzhuNi9CqTivLIbMzi27Cg9XEm6cD3T0XUudq8CPK8Zx2BH6eJCVpwrQZfssnFeOd7nLG7xQ+niyvY0frbXofQIe179biNEseGYHVFRmqc0pLnscSivH9v1VW28r7kZ+PjHrb0Ox+yCYzcmUVGas/DClN1215uLXXnl9WP7XFcb71u4ML2fqaxR1dgfs2bNUiJntLSomvXsT3+0tHg/xqFDqp/+tGpfX3TlrKRa2ZubvR9j2TLVXM58tcH28cKoVj9Bz3Wt87x0qXnftV7Tz2eqAgA9msA1OwsPtuwovWy1IGxk4SXVKiwvez5vvi5cWP9v681b9Fu/1Y5XXE8ySl7X2mxtBSZOHH6Pfs91rfNcmjy1aBEwdiwzOx3CYEfpZSvAhMnCsz3R3W+AKZb92DHz86lT5usjj9QvR70g77d+gwZeGzctXtfaHBgw5RsY8HeuvZzn0ikVDz8MzJ3LjF2HMNhR+kSxkkrQKQ2252YFCeCHD5uWxA03AE1N3stRLchPmhSsfoMG3jA3LV4+C+Xn9p13/J/rIOfZhWkyNCyJvlOO2VEohw6p3nqralubGQdpa1O97bb4x8uKSsdqgo5V2Rg3DFKO7m7V5ctVt241X7u7w9Vvd7fqokWqN9yg2tRU++9tjZX6LWvQ8VUb5zkkcMyOY3bUQFyb1G3jDt5GCzFIOSqtZhKmftevN114EyaYllatv+/tNcu95XLB33OQslZrSdbrTmVLLdU4qZzSyaVJ3TYmutsI4OXlKF68g8zxClu/Xv6+o8NsuDo0ZMb3gt60eC1r+Vy41avNo7j2Zb0J4a4saECBcFI5hcPJzN7Uqqfi7846y4yX2Vogefly4IEHgCVL3FvNo9ok7FxueKzPtmoLif/whyZhpZxDC0AXcVJ5cFa6MUXkuyJyRES22zgeJSBoRlyaJ2LHqVY9FX83cWL1BZJLz0+9cxXHVkhhMyirddsePGj/tYqqtZ737uUC0A3A1pjdwwCut3QsSoLfoOXa3nKuqlVPfuqw9PzUO1dxrN4f9ibHT7etzRuqSuNuro0BUzRsZboAmAhgu5fnMhvTIUEz4lzLiHRVtXraulX14x832Yu16tDLCiCVzlVUmYM2MiiLKmWC2nytQ4dMHc+eXf9zWa8sjgCzMQM/YsvGFJHFItIjIj39/f1xvSzVE7QVwLthb6rV0wMPAC++aBI0atVhby8wZcrwz/n88GTtWueqUgvGRnegzVZjvX3twr7WypXACy+YOqjXKnR1jz2yJrZgp6prVbVLVbva29vjelmqJ0zQYip2ZeVBpbSeit2Wxa7LnTvN16GhM+uwtRUYPRp4/fXhfzt1yjzqnatKF28b3YFR3uSU11tHhwnsx46Z3QOCrHZSxG52stVEBLsx0yslXTipUWshZD/dv8Xn5nLD3XdnnaU6fry/c2Wz61E1us9LpXqbNMmUdf5876916JB5Xj4//F7zedV581LfzQ52YwaPUdYOxGBHLopzNwKvQcXreNqhQ6odHaoi4cbeXB9f9bozgZ8AvXTpmX/rwi4MITHYBX/YmnrwfQC/BjBNRA6IyF02jksUmo2uO69jXV7HmLx0//b1AbNmma+dneG6il0fX61Ub/PmmTlwQcfrDh82cxZvucU8Jk1iN3uDs7KCiqp+3sZxKGKNNAG83moZftRbWaPIa1CptxJHedl37ABmzgw3ydmlFWfKVaq3Cy4w7bGgAZoJJlSGa2M2ElcngEexs7WNrMEgcwnDJu3UWlkkzDw5v9mGce82XqneosoopcaURN8px+xiVm1MRMSNcZuodrYOO9csibGu4muWJlcAZieBOPk5J3GOi7q0C3oCwDG7ZMfsyHGVWjnFuVtJtvKiXoUlbCvL71iXjVZH8TVPnTJp9yLA9OnAH/8Y/Jh+BDkncfQYcMUeConBrhGUXrQBM2/p9ddNmyHJi0bUy1rZmCjsJ2DauugfPmwWcd6yBVi2DJg6Nb4xqGrJIjNnnvne4wxAcSyBFgd2wyYnieYkuzETUJwbtWmT6pQpw91kSaehO7Ahpm/l3Xa257ElrfycTJ9e+dzE3c2bxs9KuZDdsGA3JrsxqY5iK2fOHODaa83l2IU0dFdXYenrA2bPBq666swylbfgstLqKCqek+LKLjt2VG65xT2lwdXPihfshk1eEhGWLbuEccWU+pYtG26hFe/Ca7Xgylsdd9wRX9JGVLy03Op9luJMXnGZpVYw2LIL/GCwI7ckfXGstZpHc3P1C1b5RX/SpPR2tZUKG8QbPHvyNBa6YRnsgj/YjUluSXouYG+vWbmjuLMAYL6fNw/Yt696t12xm3j2bDPxfO/eM7uriskJ27alJ0mhvOvw+ee9nZ9a3XaNmqSR5m7YLEgiwrJll0JRt7hcSvKota6il267aq2/YiunWsKHy/yeHy/1kKb37wiwZRf4wWCXRkl09UV9gXJpseLubtMNecst5jFpkr9xzfLuqvIJ4tUCRtJduLUEOT9e6yGtWasJYLAL/mA3ZhrF2dUXVxaZS4sVr19vujPXrTOP3l5/89zKu6uuu85ka5bXWWvr6VmbSXfh1lLp/OTzwIIF1bvjqtVDVrJWKV2SiLBs2QUUVVdfrRZFnC2uLGeJFls5xdZNPj/cUk6yC9dPa9JGEk4W5solCGzZsWXXEKKaz1WtRVHcJaGpKZ6dqW2seFLpuC4otnKuvtos/3X11cNJCknO0/PTmvSShFMPkzQoKUlEWLbsQrB5Z1yvRVEcp5s0yX6Lq3RjUtt39+Xjiy6PhRXF3eKpdu6bm+vXlUvjqw0GbNkl27ITketFZLeI7BGRr9o4JlVh8864WotC9fRxur17zZ387NnhWlxFra3A6NGmBaY11uf020KrNr44bpy7Y2FFcbd4qp37hQvr15VL46tEXoWNlgDyAH4L4CIAIwFsA9BZ629S1bJLQ6sgqNLWVWmLIso792otilzuzOP7zQAtL3etyeF0emuyXqZouSyPrzoMbNkl2rK7EsAeVe1V1ZMAHgfwOQvHdYPLGXJhrVxpWk+dnae3KKK8c//1r4H29tMnbQPA7bcPHz9oBmh5uUXMVkbF1kvxNRcuDP8+sqC0NbloETB2rPdxQ1vjq0QxsRHsxgD4XcnPBwr/lm5ZXri19L0BZqHfmTOB7353+KIVVbfa2rVAf3/t/drCJGyUlnvZMuD9982WRoB5TQB45JHsnMswSgPWww8Dc+eya5Iyy0awkwr/pmc8SWSxiPSISE9/f7+Fl41Y1layL+Xlvdm+cy8PsIAJPrncmfu1hWlZlpf7ox81rZYbbjBZpdXebyOpNhZq4wbHxUxYItgJdgcAjCv5eSyAQ+VPUtW1qtqlql3t7e0WXjZiQS+4afjP7ue92Xo/1QLsgQOVA6mtluX69abVMmGCaaGz1VK9az7MDU7xc3L33dnt9qdUsxHsXgQwRUQmichIAAsBPGXhuMkLcsFNyxif1/dm6/34vXmw3bLk/K5ou+bHjQN++UvTRZy1bn/KBDEJPiEPInIjgH+Aycz8rqp+u9bzu7q6tKenJ/TrOqW11VzAy7W0AMePx1+esKJ4P/PmmaC3eLEZu+vrcyOxoTh5ft26bLf2+vqAFSuADRvMOGZbm9nh4b77gr/vap8TwLTcwxybziAiW1S1K+lypJGVeXaq+rSqTlXVyfUCXSgudxFmbYwvivfjagZfWlrjYUWRZVv+OQGGE4+S6ip2+TpBiUnXcmEuX5SyNtE2a++nkiDdemm/kNruzi39nOQKl5N580wmbFJ15PJ1ghKTjmCXlmkAWRsXytr7KRek9Zr2C2kUrevi5+Sll4Dly4HBwXha7uU3Ho1480LeJTGT3fcKKq6vxZflVVbiklQdel2T0qXNZcmotAaq3+tEyjaSBVdQSXQFlei53qWW9rt9F8RVh+V38l5br1kbk02zai24iy7yfp1IS28RWZOOYAe42aXG/zDhea1DW91N5UHVa7ee6zdcjaS31ywDV1R648GbF6omieZkqhaCrsX17tU08FqHYbubbHRDcvHj5FU7j/m8/2OlcCNZsBsz492YruLdfnj16tBW69nrnXytFqSrUycaSfE8FjM/W1pMK++66/wfy8XeIopMOoOdSxlU/A9Tm5dzVasObXU3eb0xcX381aXPfhKK5xEw5/HkSWDOHODpp/0fizcvjSWJ5mTobsyUZVBFIi0ZoDbOla3uplrdkGnJtuRnv6G7k8FuzMCPdAU7Vy5ILgQa1y96Ns+V14tbmPPi+virK599ShSDXfBHuroxXcmgSrKrKy0ZoDbPldfupjDnxfXxV1c++0Qpla5gl/QFyYVAk5aLXpznytZ5cXn8NenPPlHKpSvYAclekFwINHFc9GwlQcR1rnp7zer9xQy9oOfF9YQFl4MxkeOaki6Ab6UXoPvvj/e1Xbm7Ll70SrfKqSbI9jWl3YGrVgUvZ1znqqMD2L3btOry+ey2epL87BOlnJX97PxK9X52ru7JVs3y5cADDwBLltQPXGnck6+lBRgYOPPfczng1Kn4y0MUIe5nF1z6ujGT5npXV1GQcSwXumn9WrDAfG0qdFIUy3zwoN3XSdP8tjSVlSgmoYKdiMwXkR0iMiQivNtwSZDA5Uo3rRfFYP7oo+bnwUHz9dixaMrs+mTzUmkqK1FMwrbstgOYB+CXFsrS2GzfjQcNXLaSIKJuXZQH83weuPFGYNEiu6/pQgauV2kqK1HMQgU7Vd2lqrttFaahRXE3Xilw1QtCQbtpy48bdeuiPJirAhMmAA8/bLdruV4L2aUuwzR2QxPFhGN2SYvybrxS4IoqCBWPO3Zs5feTy9kPCHGk4tdrIbvUZZimbmiimNXNxhSRZwBU+t9yj6o+WXjOcwBWqGrVFEsRWQxgMQCMHz9+1v79+4OWOVv6+oAVK4ANG8x4U1ubmTN2333+L1K1phlElWlZ7bhFbW3AmDHAnj0mIIWZypCUShm4P/2pm5mracsWJl+YjRmCjTXHADwHoMvr8zOzn50tXhc6rrf2Y631MqNa+7HScadMURWpvJZjXOs5Rr1+qa36dGGdVUoNcG3MBlkb0w+XxlLq8dodV63LzEtXqNcuLr/1Vum4g4PAsmXApk1mr7F83jzXzxhS2PMX95hh0C5Dl7pBibIsTKQE0A3gAIABAIcB/JuXv4ulZef6rgB+1Fvxvl4ro9h6uOGG+rsHBKm3WrsSBN2eJ+j5i3N3gDBbzXAXAwoAbNkFj1dJvGikwS6LFxEvXWa1goqXwBGk3rx0wfkNCGHPn+tb9RSlpZzkFAa74I/sdWNmMf3aS5dZpa5QP5meQerNSxec36kMYc9fWjIS01JOoozIXrDL6kWk3rhepaDiJ3D4qbcop0vYOH9exkBdGNPlLgZEsUnfrgde+NkVIC2CrHjvN3B4qbe+PuDSS4E33jBTJY4fP326hA1hz5+XurK1s0MY3MWAKDbc9cA1frfkqfd82/Ouli83Lbmilhbg5Elvuyq4IO6dHYJssURUBefZBZe9bsy4Be0OK/270u/9pqLXe76tXRpKuy5LnThh/j0tXXBxj+lyagGRG5LIisnUpPJqmY5eJoCLqHZ0qC5aVDkDsVYWYtxZp4cOmUzKfH74tfJ51Xnz0pdBGHQ6hB9ZzAqmxIHZmMzGjEStVlu9JA0vE8BVzWs88siZx6/X4oi7hdLRAVxwwekbop46Zf4tbd1zcSSGZDErmCjFspmgYkutJIbe3jPXtLz+euDHPzbBrGj1avMojgn19prFkoeGqr9uc3P9ZJIksk4PHwYmTQI+9jHz84svpqf7slQciSFZzQomSqn0t+xqtb6Kv9u2zd+4WtDlt3YXdjuaMqX6HX1Hh/m5XHFJrenTgRde8NbiiDt1vTidYd068+jtjWeRYRemCQTBqQVE7kii79TqmF2t1UGKv5s+fXh8zMv4ktfVLYqrg4wcWX3MLZdTveOO08fvuruHy1R83vz5/pecahRZWvqNKARwzC7wI71TD2qlkAO1t53xkma+bJlJ1R85sn5qffk2PbkcMHmyaRGuX2+2g9m///RjcCuW+uKeJkDkOE49CC693ZiVEgDmzQNmzjTdRrfeWn1FjxMn6q/24acLqrxLEwDmzAHmzjXBbe/eM7tDbU0JcInt7kYmeRCRJekNdqUBprnZtKh27AB+8xvggQfM7wYGTCurVFOTtwum32BUKTgGvVjbCBpJjHPZnlPGJA8isiS9wQ4YDjDvv29+3r17uAW1Zo1pRV1zDXDeeeb3+bzZa63eBdNLoCh/TqXgGPRibSNoxDmZOcq1MpnkQUQ2JDFQaC1BpdrEXeD0hBK/28x4SYjwmjTh57VtTEROYjIzt6shigWYoBL4ke5gV36RLa7qIRIsc89LoIgymNgIGkkFnjhWJSFqcAx2wR+hujFF5F4ReVVEXhaRH4vIuXbamx6VdhMWx+bmzTOZlEG6u7yMsUWZNGFjjKqjw3TXHjvmbXK6LexuJCKHhV1BZROAu1V1UES+A+BuAH8Tvlg+VNoOJuiqGF6CTdRJEza2J/rVr8zXm24C2tvj2eIoru1quIsAEQVgbZ6diHQD+AtVrbA8yOmc3uLHy/y3JObIebnIN8K8tOXLTbZtWrYUIrKI8+yCsxns/hXAOlX9l3rPdTrYucrLRb58cnvppqppbwU1QiAnqoPBLri6Y3Yi8oyIbK/w+FzJc+4BMAjgsRrHWSwiPSLS09/fb6f0cUpqfUY/af1ZnpfGCeZEFELdYKeqc1T1sgqPJwFARBYBmAvgNq3RTFTVtarapapd7e3t9t6BbdWCWlKbcPq9yLuQKBLFjUGWAzkRRS5sNub1MAkpN6nqMTtFSlh5UItywrQXfi/yLixDFtWNgQuBnIhSKdTfB2++AAAJ70lEQVSYnYjsAdAM4O3CP21W1aX1/s6JMbvyhI9qY0LNzcDNNyc7DpaWRaM5rkYUKY7ZBReqZaeqF6vqOFWdWXjUDXTOKG99VOsu3Lcvue6zYnfgqlXJt9a84LgaETkq3WtjBlGtW/Kii6oHtaS6z1auBJ5/HrjiinR02XFcjYgc1XjBrlbro1pQK46DjRoFbN8e/fyu0oCsalp4HR3xjROGwXE1InJQ2BVU0qdW66PeKiClXZ9RBrzeXmDsWNPyLFXch8/l8a+4VlIhIvKh8Vp2gP/Wh62MTK8p+R0dprVZqt4+fEnNAyQiSoHsBTsvF/1K6fm1/s5W4oWflPz33gOmTzdB1cs+fEnNAyQiSoHsBbugF/1afxc28SJIy3D9emDqVLODw5YtZrmwSoE46XmAREQpYG1tTD8imWcXdI6X178LM9ctyjUrs7weJhGdhvPsgstOyy5oV6PXvwuzMkmUKflM9yciqis7wS7oRT+uYBFlSj7T/YNjYg9RQ8hONyYQvKsxLctxkX3cH49ShN2YwWUr2BF5xXU8KYUY7ILLTjcmkR9cx5OooTDYUWNiYg9RQ2Gwo8bFxB6ihtF4a2MSFXEdT6KGwZZd3JjqTkQUu1DBTkRWisjLIrJVRDaKyGhbBcssrmFJRBS7sC27e1X1clWdCeAnAP7WQpniE2cri2tYEhElJlSwU9U/lvx4FoD4J+2FEWcri6nuRESJCZ2gIiLfBnAHgKMA/l3oEsWhfELx6tXmEeWEYqa6ExElpm7LTkSeEZHtFR6fAwBVvUdVxwF4DMBf1jjOYhHpEZGe/v5+e+8giKRaWUx1JyJKRN2WnarO8Xis/wPg/wL4epXjrAWwFjDLhXktYCSSamUx1Z2IKBFhszGnlPx4E4BXwxUnRmxlERE1jLBjdv9NRKYBGAKwH8DS8EWKSVytrL4+YOFCYN06js+xLogoIWGzMW9W1csK0w/+vaoetFWwzOC8umGsCyJKCLf4iQq3kBnGuiCyglv8BMflwqLCeXXDWBdElDAGu6hwXt0w1gURJYzBLkrM+BzGuiCiBHHMjogoJThmFxxbdkRElHkMdkRElHnpDXbcBJWIiDxKb7DjBGUiIvIofcGOm6ASEZFP6Qt2YScos/uTiKjhpC/YhZ2gzO5PIqKGk75gBwSboMzuTyKihhV2i59kBNmep7cXWLEC2LABOHbMdH92dwP33RdNGYmIyBnpbNkFwfUZiYgaVuMEO4DrMxIRNSgr3ZgisgLAvQDaVfUtG8eMRFy7kxMRkVNCt+xEZByAPwfwRvjiEBER2WejG/N/APhrAPFvn0BERORBqGAnIjcBOKiq2yyVh4iIyLq6Y3Yi8gyASimL9wD4GoDrvLyQiCwGsBgAxo8f76OIRERE4QTevFVEPgLgZwCOFf5pLIBDAK5U1Zppjty8lYjIP27eGlzgbExVfQXAqOLPIrIPQJfT2ZhERNSQGmueHRERNSRrwU5VJ6a2VcedEIiIMo0tO4A7IRARZVxjBzvuhEBE1BAaO9iF3QiWiIhSobGDHXdCICJqCI0d7ADuhEBE1ADSuXmrTdwJgYgo89iyIyKizGOwIyKizGOwIyKizGOwIyKizGOwIyKizGOwIyKizAu8n12oFxXpB7A/4J+fD8DFBadZLn9YLn9YLn9cLRcQrmwTVLXdZmEaRSLBLgwR6XFx80KWyx+Wyx+Wyx9XywW4XbYsYzcmERFlHoMdERFlXhqD3dqkC1AFy+UPy+UPy+WPq+UC3C5bZqVuzI6IiMivNLbsiIiIfHE+2InIvSLyqoi8LCI/FpFzqzzvehHZLSJ7ROSrMZRrvojsEJEhEamaWSUi+0TkFRHZKiI9DpUr7vr6kIhsEpHXC1/Pq/K8U4W62ioiT0VYnprvX0SaRWRd4fcviMjEqMris1x3ikh/SR19MaZyfVdEjojI9iq/FxH5X4VyvywiVzhSrmtE5GhJff1tDGUaJyI/F5Fdhf+LX67wnETqq6GpqtMPANcBaCp8/x0A36nwnDyA3wK4CMBIANsAdEZcrksBTAPwHICuGs/bB+D8GOurbrkSqq//DuCrhe+/Wuk8Fn73Xgx1VPf9A1gOYE3h+4UA1jlSrjsB/FNcn6eS1/00gCsAbK/y+xsB/BSAAJgN4AVHynUNgJ/EXFcdAK4ofH82gNcqnMdE6quRH8637FR1o6oOFn7cDGBshaddCWCPqvaq6kkAjwP4XMTl2qWqu6N8jSA8liv2+ioc/5HC948A+A8Rv14tXt5/aXl/COBaEREHypUIVf0lgN/XeMrnADyqxmYA54pIhwPlip2q9qnqS4Xv3wWwC8CYsqclUl+NzPlgV+Y/wdwNlRsD4HclPx/AmR+upCiAjSKyRUQWJ12YgiTq6wJV7QPMxQDAqCrPaxGRHhHZLCJRBUQv7/9PzyncbB0F8OGIyuOnXABwc6Hr64ciMi7iMnnl8v/Bq0Rkm4j8VESmx/nChe7vjwJ4oexXLtdXJjmxU7mIPAPgwgq/ukdVnyw85x4AgwAeq3SICv8WOs3US7k8+ISqHhKRUQA2icirhbvRJMsVe335OMz4Qn1dBOBZEXlFVX8btmxlvLz/SOqoDi+v+a8Avq+qAyKyFKb1+ZmIy+VFEvXlxUswS2y9JyI3AtgAYEocLywiHwDwIwBfUdU/lv+6wp+4UF+Z5USwU9U5tX4vIosAzAVwrapW+kAcAFB6hzsWwKGoy+XxGIcKX4+IyI9huqpCBTsL5Yq9vkTksIh0qGpfobvmSJVjFOurV0Seg7krth3svLz/4nMOiEgTgA8i+u6yuuVS1bdLfvzfMOPYLojkMxVWaZBR1adFZJWInK+qka6bKSIjYALdY6q6vsJTnKyvLHO+G1NErgfwNwBuUtVjVZ72IoApIjJJREbCJBRElsnnlYicJSJnF7+HSbapmDUWsyTq6ykAiwrfLwJwRgtURM4TkebC9+cD+ASAnRGUxcv7Ly3vXwB4tsqNVqzlKhvXuQlmPMgFTwG4o5BlOBvA0WK3dZJE5MLiWKuIXAlzzXu79l+Ffk0B8M8Adqnq31d5mpP1lWlJZ8jUewDYA9O3vbXwKGbIjQbwdMnzboTJevotTHde1OXqhrk7GwBwGMC/lZcLJqtuW+Gxw5VyJVRfHwbwMwCvF75+qPDvXQAeLHz/ZwBeKdTXKwDuirA8Z7x/AN+EuakCgBYATxQ+f78BcFHUdeSxXH9X+CxtA/BzAJfEVK7vA+gD8H7h83UXgKUAlhZ+LwDuL5T7FdTIUI65XH9ZUl+bAfxZDGX6JEyX5Msl160bXaivRn5wBRUiIso857sxiYiIwmKwIyKizGOwIyKizGOwIyKizGOwIyKizGOwIyKizGOwIyKizGOwIyKizPv/Teb2CEAixToAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Solução\n",
    "\n",
    "# Imports\n",
    "import numpy as np\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')\n",
    "\n",
    "# Variáveis x e y\n",
    "num_points = 200\n",
    "X_entrada = np.linspace(-2, 2, num_points)\n",
    "y_saida = X_entrada + np.random.normal(0, 1, num_points)\n",
    "\n",
    "# Print shapes\n",
    "print(np.shape(X_entrada))\n",
    "print(np.shape(y_saida))\n",
    "\n",
    "# Plot\n",
    "plt.plot(X_entrada, y_saida, 'r*', label='Dados')\n",
    "plt.legend(loc=\"upper left\", bbox_to_anchor=(1,1))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-2.        , -1.9798995 , -1.95979899, -1.93969849, -1.91959799,\n",
       "       -1.89949749, -1.87939698, -1.85929648, -1.83919598, -1.81909548,\n",
       "       -1.79899497, -1.77889447, -1.75879397, -1.73869347, -1.71859296,\n",
       "       -1.69849246, -1.67839196, -1.65829146, -1.63819095, -1.61809045,\n",
       "       -1.59798995, -1.57788945, -1.55778894, -1.53768844, -1.51758794,\n",
       "       -1.49748744, -1.47738693, -1.45728643, -1.43718593, -1.41708543,\n",
       "       -1.39698492, -1.37688442, -1.35678392, -1.33668342, -1.31658291,\n",
       "       -1.29648241, -1.27638191, -1.25628141, -1.2361809 , -1.2160804 ,\n",
       "       -1.1959799 , -1.1758794 , -1.15577889, -1.13567839, -1.11557789,\n",
       "       -1.09547739, -1.07537688, -1.05527638, -1.03517588, -1.01507538,\n",
       "       -0.99497487, -0.97487437, -0.95477387, -0.93467337, -0.91457286,\n",
       "       -0.89447236, -0.87437186, -0.85427136, -0.83417085, -0.81407035,\n",
       "       -0.79396985, -0.77386935, -0.75376884, -0.73366834, -0.71356784,\n",
       "       -0.69346734, -0.67336683, -0.65326633, -0.63316583, -0.61306533,\n",
       "       -0.59296482, -0.57286432, -0.55276382, -0.53266332, -0.51256281,\n",
       "       -0.49246231, -0.47236181, -0.45226131, -0.4321608 , -0.4120603 ,\n",
       "       -0.3919598 , -0.3718593 , -0.35175879, -0.33165829, -0.31155779,\n",
       "       -0.29145729, -0.27135678, -0.25125628, -0.23115578, -0.21105528,\n",
       "       -0.19095477, -0.17085427, -0.15075377, -0.13065327, -0.11055276,\n",
       "       -0.09045226, -0.07035176, -0.05025126, -0.03015075, -0.01005025,\n",
       "        0.01005025,  0.03015075,  0.05025126,  0.07035176,  0.09045226,\n",
       "        0.11055276,  0.13065327,  0.15075377,  0.17085427,  0.19095477,\n",
       "        0.21105528,  0.23115578,  0.25125628,  0.27135678,  0.29145729,\n",
       "        0.31155779,  0.33165829,  0.35175879,  0.3718593 ,  0.3919598 ,\n",
       "        0.4120603 ,  0.4321608 ,  0.45226131,  0.47236181,  0.49246231,\n",
       "        0.51256281,  0.53266332,  0.55276382,  0.57286432,  0.59296482,\n",
       "        0.61306533,  0.63316583,  0.65326633,  0.67336683,  0.69346734,\n",
       "        0.71356784,  0.73366834,  0.75376884,  0.77386935,  0.79396985,\n",
       "        0.81407035,  0.83417085,  0.85427136,  0.87437186,  0.89447236,\n",
       "        0.91457286,  0.93467337,  0.95477387,  0.97487437,  0.99497487,\n",
       "        1.01507538,  1.03517588,  1.05527638,  1.07537688,  1.09547739,\n",
       "        1.11557789,  1.13567839,  1.15577889,  1.1758794 ,  1.1959799 ,\n",
       "        1.2160804 ,  1.2361809 ,  1.25628141,  1.27638191,  1.29648241,\n",
       "        1.31658291,  1.33668342,  1.35678392,  1.37688442,  1.39698492,\n",
       "        1.41708543,  1.43718593,  1.45728643,  1.47738693,  1.49748744,\n",
       "        1.51758794,  1.53768844,  1.55778894,  1.57788945,  1.59798995,\n",
       "        1.61809045,  1.63819095,  1.65829146,  1.67839196,  1.69849246,\n",
       "        1.71859296,  1.73869347,  1.75879397,  1.77889447,  1.79899497,\n",
       "        1.81909548,  1.83919598,  1.85929648,  1.87939698,  1.89949749,\n",
       "        1.91959799,  1.93969849,  1.95979899,  1.9798995 ,  2.        ])"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "X_entrada"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([-2.60469233e-01, -6.12297797e-01, -3.61560726e+00, -1.64811686e+00,\n",
       "       -1.81052008e+00, -1.76981567e+00, -2.48438321e+00, -1.19245864e+00,\n",
       "       -1.62415259e+00, -1.63206332e+00, -1.66968132e+00, -3.73680887e+00,\n",
       "       -2.23721829e+00, -2.99604188e+00, -4.28852755e+00, -2.49103995e+00,\n",
       "       -2.17479730e+00, -1.84656259e+00, -1.54698561e+00, -2.79644710e-03,\n",
       "       -2.11923426e+00, -2.37378742e+00, -6.89579407e-01, -1.83179516e+00,\n",
       "       -1.46678268e+00, -4.05586300e-01, -1.58728072e+00, -1.38862351e-01,\n",
       "       -1.50618696e+00, -3.25594816e+00, -1.69774613e+00, -2.44802278e+00,\n",
       "       -2.91728178e+00, -1.46264372e+00, -1.53675938e+00, -1.46895612e+00,\n",
       "       -9.83063432e-01, -1.46019283e+00, -1.78560361e+00, -2.28464552e+00,\n",
       "       -2.30512483e+00, -2.03008859e+00, -1.13992887e+00,  2.57918124e-01,\n",
       "       -6.43484331e-01, -1.43510430e+00, -2.43811666e+00, -1.13787851e+00,\n",
       "       -1.97080721e+00, -1.79656286e+00, -8.26145323e-01, -1.23487049e+00,\n",
       "        7.05223696e-01, -1.05539584e+00,  5.92288255e-01,  2.09326761e-02,\n",
       "        1.56774091e-01,  1.01019586e+00, -1.21702696e+00, -7.88567622e-01,\n",
       "        1.30302347e-01, -1.47676569e-01,  3.01690254e-01, -1.39037409e+00,\n",
       "       -1.40759719e+00, -1.26889108e+00,  1.04422646e+00, -1.05073311e+00,\n",
       "       -1.77638582e+00, -2.17552032e+00, -1.00234630e+00, -6.08887628e-01,\n",
       "       -1.07643597e+00, -3.23298567e+00, -4.94756908e-02,  9.10012321e-02,\n",
       "        1.42186116e-01,  6.44379825e-01, -4.01752844e-01,  3.56703927e-01,\n",
       "       -1.19975739e+00,  6.09911402e-02, -1.63693770e+00, -4.86646944e-01,\n",
       "       -2.00877392e+00,  1.70539051e+00,  1.28410096e+00, -8.79455677e-01,\n",
       "       -8.06938461e-01, -6.93996699e-01,  1.07433478e-01, -1.14189356e+00,\n",
       "        7.53435662e-01,  1.05600143e-01, -3.59721853e-03, -4.38299500e-01,\n",
       "       -1.35307789e+00, -2.65274256e+00, -2.40344627e-01,  1.41691184e+00,\n",
       "       -5.57863607e-01,  4.39680123e-01,  4.36064542e-01,  7.30060349e-01,\n",
       "       -9.31549956e-01, -3.23004973e-01,  4.89612259e-01,  1.65673688e+00,\n",
       "       -1.04683019e+00,  1.35051812e+00,  5.59885068e-01,  8.42042240e-01,\n",
       "        4.11855534e-01, -1.43290037e+00,  1.06587176e+00,  3.09518714e-01,\n",
       "        3.18438982e-01,  9.16183641e-01, -1.16026812e+00,  1.05169449e+00,\n",
       "        2.53011388e-01,  1.25622227e+00,  7.32402263e-01,  7.86298149e-01,\n",
       "        9.18679091e-01,  1.02395888e+00, -6.14667122e-01,  1.00275207e+00,\n",
       "        3.63087567e+00,  2.25002485e+00,  1.42596040e+00,  1.31146772e+00,\n",
       "       -9.13642693e-01,  1.09801351e+00,  1.89028933e+00,  2.10021883e+00,\n",
       "        8.91419204e-01, -1.30324906e-01,  2.29377410e+00,  1.80587101e-01,\n",
       "        1.37985490e+00,  3.29686065e+00,  1.02193510e-01,  6.61020357e-01,\n",
       "        1.27047235e+00,  5.27795936e-01,  2.80564901e+00,  3.80766092e-01,\n",
       "        9.96252002e-01,  2.19216314e-01,  6.27271665e-01, -1.49988787e-01,\n",
       "       -1.01023504e-01,  1.37301388e+00,  1.30265762e+00,  4.14008386e-01,\n",
       "        3.63752855e+00,  2.37053777e+00,  3.21754301e-01,  1.75693998e+00,\n",
       "        1.11809512e+00,  1.23126090e+00,  8.32285559e-01,  3.65004330e+00,\n",
       "        1.87470164e+00,  8.60843242e-01,  1.01955925e+00,  2.44562310e+00,\n",
       "        6.30355233e-01, -4.39054280e-01, -3.48022848e-01,  1.81933953e+00,\n",
       "        1.29032405e+00,  1.91835440e+00, -2.94654558e-01,  9.13058000e-01,\n",
       "        9.47908741e-01,  1.43012995e+00,  2.08997232e+00,  1.18364243e+00,\n",
       "        1.42423360e+00, -3.62701776e-01,  2.76829055e+00,  1.14282020e+00,\n",
       "        1.66979245e+00,  2.53586974e+00,  2.24736001e+00,  2.20719888e+00,\n",
       "        1.93102196e+00,  1.32521921e+00,  1.99872685e+00,  3.06241692e+00,\n",
       "        2.83870505e+00,  3.56208972e+00,  1.48282445e+00,  2.30121476e+00,\n",
       "        2.11349992e+00,  1.41186607e+00,  3.55161706e+00,  5.32441660e-01])"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "y_saida"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(200, 2)\n",
      "(200, 1)\n",
      "(2, 2)\n",
      "(2, 1)\n",
      "\n",
      "Coeficientes da Regressão:\n",
      "[[1.0787248699273968], [0.011504377438197004]]\n",
      "\n",
      " [-2.14594536e+00 -2.12426245e+00 -2.10257954e+00 -2.08089663e+00\n",
      " -2.05921371e+00 -2.03753080e+00 -2.01584789e+00 -1.99416498e+00\n",
      " -1.97248207e+00 -1.95079915e+00 -1.92911624e+00 -1.90743333e+00\n",
      " -1.88575042e+00 -1.86406751e+00 -1.84238459e+00 -1.82070168e+00\n",
      " -1.79901877e+00 -1.77733586e+00 -1.75565295e+00 -1.73397004e+00\n",
      " -1.71228712e+00 -1.69060421e+00 -1.66892130e+00 -1.64723839e+00\n",
      " -1.62555548e+00 -1.60387256e+00 -1.58218965e+00 -1.56050674e+00\n",
      " -1.53882383e+00 -1.51714092e+00 -1.49545800e+00 -1.47377509e+00\n",
      " -1.45209218e+00 -1.43040927e+00 -1.40872636e+00 -1.38704344e+00\n",
      " -1.36536053e+00 -1.34367762e+00 -1.32199471e+00 -1.30031180e+00\n",
      " -1.27862888e+00 -1.25694597e+00 -1.23526306e+00 -1.21358015e+00\n",
      " -1.19189724e+00 -1.17021432e+00 -1.14853141e+00 -1.12684850e+00\n",
      " -1.10516559e+00 -1.08348268e+00 -1.06179976e+00 -1.04011685e+00\n",
      " -1.01843394e+00 -9.96751029e-01 -9.75068117e-01 -9.53385205e-01\n",
      " -9.31702293e-01 -9.10019381e-01 -8.88336469e-01 -8.66653557e-01\n",
      " -8.44970645e-01 -8.23287733e-01 -8.01604821e-01 -7.79921909e-01\n",
      " -7.58238997e-01 -7.36556085e-01 -7.14873173e-01 -6.93190261e-01\n",
      " -6.71507349e-01 -6.49824437e-01 -6.28141525e-01 -6.06458613e-01\n",
      " -5.84775701e-01 -5.63092789e-01 -5.41409877e-01 -5.19726966e-01\n",
      " -4.98044054e-01 -4.76361142e-01 -4.54678230e-01 -4.32995318e-01\n",
      " -4.11312406e-01 -3.89629494e-01 -3.67946582e-01 -3.46263670e-01\n",
      " -3.24580758e-01 -3.02897846e-01 -2.81214934e-01 -2.59532022e-01\n",
      " -2.37849110e-01 -2.16166198e-01 -1.94483286e-01 -1.72800374e-01\n",
      " -1.51117462e-01 -1.29434550e-01 -1.07751638e-01 -8.60687264e-02\n",
      " -6.43858144e-02 -4.27029025e-02 -2.10199905e-02  6.62921459e-04\n",
      "  2.23458334e-02  4.40287454e-02  6.57116573e-02  8.73945693e-02\n",
      "  1.09077481e-01  1.30760393e-01  1.52443305e-01  1.74126217e-01\n",
      "  1.95809129e-01  2.17492041e-01  2.39174953e-01  2.60857865e-01\n",
      "  2.82540777e-01  3.04223689e-01  3.25906601e-01  3.47589513e-01\n",
      "  3.69272425e-01  3.90955337e-01  4.12638249e-01  4.34321161e-01\n",
      "  4.56004073e-01  4.77686985e-01  4.99369897e-01  5.21052808e-01\n",
      "  5.42735720e-01  5.64418632e-01  5.86101544e-01  6.07784456e-01\n",
      "  6.29467368e-01  6.51150280e-01  6.72833192e-01  6.94516104e-01\n",
      "  7.16199016e-01  7.37881928e-01  7.59564840e-01  7.81247752e-01\n",
      "  8.02930664e-01  8.24613576e-01  8.46296488e-01  8.67979400e-01\n",
      "  8.89662312e-01  9.11345224e-01  9.33028136e-01  9.54711048e-01\n",
      "  9.76393960e-01  9.98076872e-01  1.01975978e+00  1.04144270e+00\n",
      "  1.06312561e+00  1.08480852e+00  1.10649143e+00  1.12817434e+00\n",
      "  1.14985726e+00  1.17154017e+00  1.19322308e+00  1.21490599e+00\n",
      "  1.23658890e+00  1.25827182e+00  1.27995473e+00  1.30163764e+00\n",
      "  1.32332055e+00  1.34500346e+00  1.36668637e+00  1.38836929e+00\n",
      "  1.41005220e+00  1.43173511e+00  1.45341802e+00  1.47510093e+00\n",
      "  1.49678385e+00  1.51846676e+00  1.54014967e+00  1.56183258e+00\n",
      "  1.58351549e+00  1.60519841e+00  1.62688132e+00  1.64856423e+00\n",
      "  1.67024714e+00  1.69193005e+00  1.71361297e+00  1.73529588e+00\n",
      "  1.75697879e+00  1.77866170e+00  1.80034461e+00  1.82202753e+00\n",
      "  1.84371044e+00  1.86539335e+00  1.88707626e+00  1.90875917e+00\n",
      "  1.93044209e+00  1.95212500e+00  1.97380791e+00  1.99549082e+00\n",
      "  2.01717373e+00  2.03885665e+00  2.06053956e+00  2.08222247e+00\n",
      "  2.10390538e+00  2.12558829e+00  2.14727121e+00  2.16895412e+00]\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<matplotlib.legend.Legend at 0xb2ca18470>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAl4AAAD8CAYAAACmaqpKAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADl0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uIDIuMi4yLCBodHRwOi8vbWF0cGxvdGxpYi5vcmcvhp/UCwAAIABJREFUeJzt3Xl4VNX5B/DvmQnZZJElYFgTIYEsbCUqBBCroKCALCIQZKkoEGpbtbgrbQUXqrX+VNa6IEoAbVFwB7WKiliDEgg7hj0BImAIJmSb8/vjZMgwzD53Zu6dfD/Pc58kkzv3nrl3yLyc8573CCkliIiIiCjwTKFuABEREVF9wcCLiIiIKEgYeBEREREFCQMvIiIioiBh4EVEREQUJAy8iIiIiIKEgRcRERFRkGgWeAkhzEKIH4UQ72t1TCIiIqJwomWP158A7NTweERERERhJUKLgwgh2gK4CcATAO51t3+LFi1kQkKCFqcmIqo3Nm/e/LOUMi7U7SAi32kSeAF4HsD9ABp5snNCQgJyc3M1OjURUf0ghDgY6jYQkX/8HmoUQgwFcEJKudnNftOEELlCiNzi4mJ/T0tERERkOFrkePUFMFwIcQDASgDXCiHetN9JSrlESpkhpcyIi2NPOREREdU/fgdeUsqHpJRtpZQJAMYB+FxKeZvfLSMiIiIKM1rleBERUQhs3ry5ZURExMsA0sHajEShZgGQX11dfUevXr1OONpB08BLSvkFgC+0PCYRETkXERHx8mWXXZYSFxd32mQyyVC3h6g+s1gsori4OPXYsWMvAxjuaB/+74iI9KOoCBgwADh2LNQtMZL0uLi4Mwy6iELPZDLJuLi4EqgeaMf7BLE9RKQlIwUpnrZ1zhzg66+Bxx8PTrvCg4lBF5F+1P57dBpfMfAiMiojBSnu2hoTAwgBLFwIWCzqqxDqcSMwUhBMRCHFwIvIaIwUpHja1oICICsLiI1VP8fGAhMmAPv3B7/NvnAXWIZ5YGY2m3t16dIltVOnTmmdO3dO/etf/9qqpqbGq2OMHj064bXXXmsaoCYS6QYDLyKjMVKQ4mlb4+OBxo2Bc+eA6Gj1tXFj4LLLgt9mb3gaWOqtd/LgwQa44orOOHRIkwlWUVFRll27du3Yt2/f9s8//3zPunXrmsyaNau1FscmCjcMvIiMxkhBijdtPX4cmDED2LRJfTVC75C7wFKvvZOPPBKPzZsb4uGHNQ+O2rRpU/3yyy8feO2111paLBbs3r07slevXp1TU1NTUlNTU9avX38JAFgsFkyaNKl9x44d06655ppOP//88/kgcM2aNY1SUlJSk5OTU8eMGZNQXl4uAGDmzJltOnbsmJacnJw6bdq0tlq3nSgYWMeLyIisQcq0acCSJWooS688bevq1XXfz58fnLb5y11gWVAAzJoFvPsuUFamArORI4Fnnw1Ne6OifoPKSnH+5+XL47B8eRwiIyUqKn7Q6jSpqamVFosFR48ejWjdunX1V199tSc2NlZu27Ytavz48Zfn5+fvfOONNy7dt29f1O7du7cfOXKkQdeuXdOmTJlysqysTEyfPj1x3bp1u7t161YxcuTIhGeeeSZu+vTpJz/88MOmBQUF+SaTCT///LNZq/YSBRMDLyIjMlKQYqS2+sJVYKm33sk9e7bhD39oi/XrL8W5cyZER1tw/fW/4MUXD2t9KinVRMvKykoxderUDjt27IgxmUw4ePBgFAB8+eWXjW699dZTERERSEhIqOrTp08pAOTl5UW3bdu2olu3bhUAMGXKlJPz589v+dBDD52IioqyjBs3rsNNN91UMnbs2BKt20wUDAy8iIj84S6w1FPvZIcOVWjUqAYVFabaXi4TGjWqQfv21VqeZseOHZFmsxlt2rSpnjVrVuuWLVtW/ec//9lvsVgQExPTy7qfEOKi51oDNnsNGjTAli1bdq5du7bxypUrmy5cuLDlpk2b9mjZbqJgYI4XESlhPvMuZFavVgFZ9+7qq22gFgrFxQ0wYUIxvvhiJyZMKMaJEw20PHxhYWHEnXfe2eF3v/vdCZPJhJKSEnN8fHyV2WzGggULmltnOw4YMKD07bffblZdXY2DBw822LRpUyMA6NGjx7mjR49G5ufnRwHAsmXLmvfv37+0pKTEdOrUKfPYsWNLFi1adHjnzp2xWrabKFjY40VEiu3MuwULQt0aCpR16346/32fPoe0OGRFRYWpS5cuqdXV1cJsNsuxY8ee/Mtf/nIcAO6+++4To0eP7vjuu+827devX2lMTIwFACZOnPjLZ5991rhz585piYmJ56688spSAIiNjZWLFi06MGbMmI41NTXo3r172axZs4pPnDgRMXTo0E4VFRUCAObOnav58ChRMAhn3bqBlJGRIXNzc4N+XiJyICZG5R7Zi44GysuD3x5ySgixWUqZYftYXl7ege7du/8cqjYR0cXy8vJadO/ePcHR7zjUSFTfGakuGBGRwTHwIqrv9DbzjogojDHwIiJjFi8lIjIgJtcTUfjX2gqEoiJg3Dhg1Sr2DhKRx/zu8RJCRAsh/ieEyBNCbBdC/E2LhhER6Zre1l8kIkPQYqixAsC1UsruAHoAGCyE6K3BcYmI9Eev6y8SkSH4HXhJ5Wztjw1qt+DXqCAiCgbOAr1IbGxsT/vH/v73v8e99NJLzV0974UXXmg+adKk9lq0Yffu3ZFJSUlp3jzHbDb36tKlS2pSUlLatdde20lv6z++9957jXr06NFl0KBBHZcuXXqpVsdt06ZN1+Tk5NTOnTun9u3bN+nQoUOapB2NHTu2w+bNm6Od/f7UqVOmzMzMpMzMzKSpU6e28/d8u3fvjhRC9HriiSdaWh+bNGlS+xdeeMHl+85To0ePTmjTpk3XLl26pKampqZ8+umnl2hxXE2S64UQZiHEFgAnAKyXUn7nYJ9pQohcIURucXGxFqclIgo+zgL1yP3331981113nQx1O1yJioqy7Nq1a8fevXu3X3rppdXPPPNMnBbHraqq0uIwGDZsWOmWLVt2rV+//qcpU6b8oslBa3355Zd7du/evaNnz55ls2fPjrf/fXW196tIrVq16mCvXr0cFAVUmjVrZtm4cePejRs37n3llVc0KYDbrFmz6sWLF7c8d+7cxetPaWDu3LlHdu3atWPu3LlHZ86c2UGLY2oSeEkpa6SUPQC0BXClECLdwT5LpJQZUsqMuDhN3ttERKFhOwt04kTgrbc4E9TOvffe23r27NmtAODKK6/snJ2d3aZr164pCQkJ6R9//HFD637Hjh1r0L9//6QOHTqkz5gxo6318QkTJrRPT09P6dSpU9o999zT2tE5vvrqq9jOnTun9ujRo8tzzz13vtejuroa06dPb5uenp6SnJyc+swzz7Rw197evXv/evTo0Ujrz4899lgr6/Ntz3/ffffFJyYmpmVmZiYNGzYs0fY13nXXXW2uuOKKznPnzm1VWFgYccMNN3RMT09PSU9PT1m3bt0lAPDBBx807NKlS2qXLl1SU1JSUk+fPm06ePBgg4yMjM7W3jfr9XF2DdasWdMoJSUlNTk5OXXMmDEJ5eXlPgcd11xzTen+/fujANVzeffdd7fu1q1bl88++6zhV199FXvFFVd0TktLS+nXr1/SwYMHG/zwww/RXbt2TbE+f/fu3ZHJycmp1muwYcOG2OrqaowePTohKSkpLTk5OfVvf/tbSwDYuHFjTPfu3bskJyenDho0qGNxcbEZALZv3x7Vv3//pLS0tJRevXp1/vHHH6MB4NVXX22alJSU1rlz59SMjIzOjtrfrFmz6n79+pXOnz//ol4ua3sAoKioKKJNmzZdAdXTOnDgwI7XXnttpzZt2nR98skn4/7617+2SklJSe3evXuX48ePX9TzOXjw4NLDhw9Hbd++PSo1NfX869+2bVtUWlpaiv3+rmg6q1FK+YsQ4gsAgwHka3lsIgohzuC7kO0s0NhY4PRpXSy1dPvtaJefD03XMExPR9mrr8Lv3onq6mqxbdu2natWrWry+OOPtx48ePAeANixY0dsXl7ejpiYGEunTp3SZ82adbxTp05Vzz333NFWrVrVVFdXIzMzs/N3330Xc9VVV12wlMLUqVMT/vnPfx666aabzk6fPv180Pb888+3aNKkSU1+fv7O8vJyccUVV3QZNmzYmS5dulQ6aRv++9//Npo6derPALB69erG+/bti966detOKSUGDhzY6aOPPmp4ySWXWN57772m27Zt21FVVSV69OiR2rNnzzLrcX755Rfz999/vxsAhg0blnjvvfcev+GGG87u3bs38oYbbkgqKCjY/o9//OOyF1544eD111//a0lJiSk2Ntby/PPPx1133XUl8+bNO1ZdXY3S0lITADi6Bl27dj03ffr0xHXr1u3u1q1bxciRIxOeeeaZuNmzZ5/w5b6sXbv20tTU1HIAKC8vN6Wnp5c///zzhRUVFaJ3796dP/jgg32tW7eu/te//tV01qxZbd5+++0DVVVVYseOHZGpqamVy5YtazZixIjTtsf89ttvY4uKihrs3bt3OwBYh3CnTJmS+Oyzzx4ePnx46d133936gQceaP3qq68evuOOOzosWbLkYNeuXSs+//zzS7Kzs9tv2rRpz9NPPx2/bt26PYmJiVWuhoFnz55dNGTIkKQ//elPHq/gsGfPnpi8vLwd5eXlps6dO6c/9thjR3fu3Llj6tSp7RYvXtzc/nquXLny0qSkpPK0tLSKRo0a1WzcuDEmMzOzfPHixS2ysrK86tn1O/ASQsQBqKoNumIADAQwz9/jEpGOhPs6jr4ElvZLLS1cqDYuteTQmDFjTgNAZmbmr/fdd9/5nqV+/fqdad68eQ0AdOrU6dxPP/0U1alTp6rXX3+92dKlS1tUV1eL4uLiBnl5edG2gdfJkyfNpaWl5ptuuuksANx+++0nP//88yYA8OmnnzbetWtX7Nq1a5sCQGlpqXnHjh3R9oGXdY3Jo0ePRqanp5eNGDHiDAB8/PHHjTds2NA4NTU1FQDKyspMu3btii4tLTUNGTLkl4YNG0oActCgQRcM/40fP/6U9ftvvvmm8d69e8/PuDh79qz59OnTpt69e5+dNWtWu1tvvfXU+PHjT3fs2NHSu3fvX6dPn55QVVVluuWWW05nZmaWA4Cja2CxWNC2bduKbt26VQDAlClTTs6fP78lVKqPxwYMGJBsMpmQkpJS9txzzx0FALPZjClTppwGgK1bt0bt3bs35tprr00GAIvFgri4uCoAGDFixKk333yz2ZNPPnnsnXfeabpq1aoC22N36dKl4vDhw1GTJ09uN2zYsJKRI0eeOXnypPnMmTMRw4cPLwWAO++88+SYMWMuLykpMf34448Nx4wZ09H6/MrKSgEAGRkZZydMmJAwevTo0xMmTLgguLM7X2WPHj1+Xbx4cTNPX39mZmZp06ZNLU2bNrU0bNiwZsyYMb8AQNeuXcu2bt16/j8vjz76aNt58+bFN2vWrOqVV145AABTpkz5+V//+leLK6+88vCaNWuafv/99zs9PS+gTY9XPIDXhRBmqKHLt6SU72twXCIKtfoSXPgSWBYUALNmAe++C5SVqZ6vkSOBZ58NbFtd0KJnKlCio6MlAERERKCmpub80FhkZOT5yVhms1lWVVWJXbt2Rb700kutNm/evDMuLq5m9OjRCefOnbsgNUZKCSEcj7BJKcU//vGPQ6NHjz7jqk3WHK+TJ0+ar7/++k5PP/10y0cfffSElBJ333130X333XdBD4p1yMyZRo0aWWzbl5ubu7M2SDvvySefPDZixIiSNWvWNMnMzEz5+OOP9wwZMuTshg0bdv/nP/9pMmXKlMQ//vGPxwcOHFjq6Bp4sr7yvn37GgwdOjQJAG6//fbi+++//6LE6i+//HJPfHz8BYlckZGRloiICGv7RadOncq3bNmyy/65EydOPD1mzJjLx40bd1oIga5du1bY/j4uLq4mPz9/xzvvvNN4wYIFLVetWtVs0aJFh4UQFzW+pqYGjRo1qt61a9cO+9/l5OQc+vzzzy9Zu3Ztkx49eqRt2bJl+2WXXVbj6DXPnj372K233trxqquuKrU+FhERIWtq1O5lZWUXvFls33cmk+n8+9NkMqG6uvr8vnPnzj3yu9/97oKgb/LkyafnzZvXeuXKlaVdu3Ytc9YmZ7SY1bhVStlTStlNSpkupWRRG6JwEe4z+PwpDcEk+4A5ffq0OSYmxtKsWbOaw4cPR3zxxRdN7Pdp0aJFTcOGDWs++eSThgCwdOnS870dgwYNKlm4cGFcRUWFAFTvzZkzZ5x+3jVv3rzmhRdeODR//vxWFRUVYsiQIWfeeOONFiUlJSYA2L9/f4OjR49GXHPNNWc/+eSTJmVlZaKkpMT06aefOp1p2K9fvzPz5s07H6ht3LgxBlD5TFdeeWX5E088caxr166/5ufnR+/ZsyeyTZs2VX/+859/vu22237+4YcfYp1dgx49epw7evRoZH5+fhQALFu2rHn//v1Lbc/dqVOnql27du3YtWvXDkdBlye6det27tSpUxHWmXwVFRUiNzc3GgDS0tIqTCYTZs+e3XrkyJGn7J9bVFQUUVNTgylTpvwyd+7co9u2bYtt3rx5TePGjWus+WuvvPJK8z59+pxt1qyZpW3btpWvvvpqU0D1rH377bfnr9W111776/PPP1/YtGnT6oKCgkj7c1n17NnzXFJSUvlnn312/r3Srl27iv/973+XAMDy5cub+nIdHImNjZUDBgwouffee9tPmTLF6wXqWbmeiJwL9+DC314ra5L9tGnAkiVqyLIeOnfunKlVq1bdrD9nZ2cf9+d4ffr0KU9PTy9LSkpKa9++fUWvXr3OOtrvlVdeOXDHHXckxMTEWK699trzvVv33HPPzwcOHIjq2rVripRSNGvWrOrDDz/8ydU5+/btW56SklL+8ssvN/39739/avv27dFXXHFFFwCIjY21LF++fP+AAQPKBg8eXJKamprWpk2bim7duv3apEkTh70dS5YsOXzHHXe0T05OTq2pqRFXXXVVaWZm5qG///3vLTdu3NjYZDLJ5OTk8ltuuaXk5ZdfbvbCCy9cFhERIWNjY2uWL1++v0uXLpWOrkFsbKxctGjRgTFjxnSsqalB9+7dy2bNmqV5qYDo6Gi5cuXKn/74xz+2Ly0tNdfU1Ijs7OzjGRkZ5wBg1KhRp+bMmdN23rx5R+2fe+DAgQZTp05NsFgsAgAef/zxIwDw2muv7c/Ozu7wxz/+0dS+ffuKFStWHACAFStWFNx5550d5s2bF19dXS1Gjhx5qk+fPuX33HNP2wMHDkRJKUW/fv3O9O7d22U3+2OPPVbUt2/fVOvPDz744PGxY8devnLlyub9+/d32fvprUmTJp366KOPmo4aNcrr4wpPui21lpGRIXNzc4N+XiLywahRKgCzDS5sk8uNLjtbva7ISKCyEpg+Xbd5bEKIzVLKDNvH8vLyDnTv3t3r/3WTb0pKSkxNmjSxlJaWmvr06dN50aJFB/v161fm/pkUTmbPnt2qpKTE/H//93+Fjn6fl5fXonv37gmOfsceLyJyLdzXcWSvFXnhtttu67B3796YiooKMW7cuJMMuuqfQYMGdTx48GDUl19+uceX5zPwIqL6zVFgyfIZ5MR7770XJgmO5Kv169e7HLZ2R5MCqkREYcVYC2BbrLk0RBR6tf8eLc5+z8CLiIyhqAgYMCCwFeKNuQB2fnFxcRMGX0ShZ7FYRHFxcRO4KCLPoUYiMoZgFHHVYW0ud6qrq+84duzYy8eOHUsH/zNNFGoWAPnV1dV3ONuBgRcRBYeveVPBLOJqwPIZvXr1OgFgeKjbQUSe4f+OiCg4fM2bCnYRV9sFsGfM4OLXRKQpBl5E9VEw8qWs/M2bCnYv1OrVanZj9+7qq7c1y4J5bYnIcBh4ERmZrx/ywZy1p0WPVah6oXy5vsaaEUlEQcbK9URGNnMmsHix59XW7fOlrAK96LWBqsNfYOZMYNEi1bv2ww+ue9mCcG0dVa4nImNhjxeREfk6fBeqRa+Nljdle32lVD1f8fGur2+4LyhORJrgrEYiI/K17EGoZu0ZbdmhggKgbVsV1No6d04FX456sAw4I5KIgo89XkRG5O2HvG2uktF6n7TiTb5WfLzqrbIVEeG+B8v22k6cCLz1FnDsGCoqgDVrgLFjgcOH/XsZRGRs7PEiMipvFne2Tfg2Wu+TVrwtwHr2LJCWBuzYAZhMQHW1+x4sm2tribkEG051w/Jr9+HfRZfhl1+AuDhg926gXTsNXg8RGZLfyfVCiHYAlgG4DKpi6xIp5f+5eg6T64mCJFTJ9HrizzUYNUr1ftkGt6tXOy0GKyWwJbo3cipHYwXG4yjaoiFKMRLvIKvBv3Hdr2vRoIHvL4XJ9UTGp8VQYzWAP0spUwD0BvB7IUSqBsclIn8x4du/a+CsppddyYiffgLmzgVSU4HfVG7C8+Ie/Mach5UYi+MxiVg2YR0GH1riV9BFROHB76FGKWURgKLa70uFEDsBtAGww99jE5Gf6lvCt6OeKK2uQVHRBQn3x9ESby00Y/nCTfgOvQEAV18N3HMPMPrbB9F82T+B6EigojK8rzkReUXT5HohRAKAngC+c/C7aUKIXCFEbnFxsZanJSJnioqAt99Wid5GSqbXujCsFhMK5szBGUtDLGt1H24wrUdrFOKPeBHnmsZj3qOlOHQI+PJLNSrZvKSgfk5gICK3NCugKoRoCOBLAE9IKV2uscEcL6Ig8bbAql7oqDBsRXQTfFxxDXKQhbUYjnOIQQL2Iws5yBIrkTajf9CuLXO8iIxPk8BLCNEAwPsAPpFSPudufwZeRAFm1KR6X9pdVKRqmLVuDXzyycV1zXwY4rNYgA0bgJwc4N9vW3D6FxNaoBhjsQoTxAr07lgMsWhhXaK9t+s5+oiBF5Hx+T3UKIQQAF4BsNOToIuIgkDvSfXOhhJ9afecOcD336s6DX7kcUkJ/PgjcN99QPv2wG9/qwKvm4aa8NGNL6JQtMVL0fehj9gEMWggcN11vi2iTUT1mhZ1vPoCmAhgmxBiS+1jD0spP9Tg2ETkC70n1TurqeVNu+17x3bUzuexWFRelau6ZjZ++glYsUIFWTt3qjqpQ4aozrLhw2tjwFH/BbLv8KxmGhGRC1wkm4zHSQ0lsuOsBlUoeTKU6Gm7i4qcL5vk5n1x/LgqKp+To/LfATUjMSsLuOUWoHlzP19ngHCokcj4WLmejMfbCuSBovcAUI8V6j1ZY9LTdnvZq3fmjDptTg7w6adATY0qzzVvnrqN7dvX7lhUBAzQ8X0lIkPjWo1kHDExgBDAwoVqOGnhQvVzTExo2uOsdAE5p/UQqJsyEZWVdWsktmoFTJ6sUsEeeADIzwe2bAHuv98m6AJ4X4kooDjUSMbhx9CSpow6Y1AvAjwEesGMxH8Dp08DLVqo4CsrC+jTR8XrFzHAfeVQI5HxcaiRjEMvCeOeDJeRcwEYApVS9V7l5KhE+aNHgUsuUbclKwsYOBDul+vhfSWiIGDgRcZiHVoK5ewyvQSA5HJG4rBhKvjymBb3Ve95f0QUcgy8yFj0kjCuhwCwnjpxQsU19jMS//QnDWYk+ntf9TLxg4h0izleRFZG7q0wcts9UFoKvPPOxTMSs7LsZiQ6Y399tL5ezvLDoqKAq67S7DzM8SIyPs5qJLIy8mw2f9ru64LUAT5eZSWwdq1Kim/Z0sMZic7YXx+t77Wzivvjxhn3PUVEgSGlDPrWq1cvSaQb0dFSqvzsC7foaM+PUVgo5dVXS1lUFLh2OuKs7VFRnh8jO1tKk0l91YIfx6upkfKLL6S8804pmzZVL6VFCyl//3spv/lGSovFywM6uz6+3mtX93nGDPW6XZ3Tm/eUAwByZQj+ZnPjxk27jT1eZFxa9dRosa5hqHrL7NtuNquv48a5f667umjeXl9nx4uOdvk0WTsj0bpG4jXXqCHFG28EPvwQKCwEXnoJyMx0UgbCVVvtr09MDJCQUPcavb3Xru6zbU2xyZOBtm31u1YmEYUMAy8yLq2CHX9ms2ld1NXbYMfa9rIy9XNNjfr6+uvu2+Eu4PT2+noZBBYUAHPnAmmdq9CzJ/D88xK/+Y2apXj8OPDmm2qGotsyEK7aan9vKypU+yoqvLvXntzn1avVhI/u3YGlS4GhQznzlYguwsCLjCcQFezdVEB3SoveMlu+BJPHj6seliFDVD0FT9vhLOBMTPTt+noQBJ44oXqv+vQBOnYEHnsMaHH2ABaJbBybeD/WrlVxmsdlIDx5L9jf29Onvb/XvtxnX99TRBTeQjG+yRwv8kthoZRZWVLGxkoJqK8TJgQ/v8rKNrfH11wpLfLMfGnHyJFSzpwp5ZYt6uvIkf5d35EjpZw8WcohQ6SMiJASkGdiWsplfRbIG645J81mdchu3aScF/GwPIh2/r1mX9rqaz6eFvfZT2COFzduht/Y40XGo7cCplr0bGjRc+ZLO2yHx+bPVz/7c31XrwaWLkVl28uxtuYmjDO9hVbl+zHp22zs2h+F++8Htm0D8vKA+w/dhfYjMwCTyffX7EtbnfUquhvmZQ8WEWmABVTJmPRUwFSLoq5aBJP27bAGEr7UkPLh+loswFdfAcuXA/9e+hROy0ZocWk1ftfmG0xo9jH6fPHUhcnx8fGqPoTFovLBfA2gPW2rfa2thQvVZl2L0V3xU70U7yUiQ2MBVfJPmBfu1Iyr62T93SWXqPwqrRaPnjkTWLwYmD49YFXUpVS9V9Y1Eo8cUS9jxAjVeeV0jURnBUdNprrcMK05W2T93/9Wyfb2dLQ4thULqBIZnyZDjUKIV4UQJ4QQ+Vocj0LA19IMRi46GkyurpP1dwkJFw/7WdneH3f3KhCTD+wUbDqBJxL/hbTO1ejZE/jnP4EePbyYkehsaPXo0Yv31apsiLNexf37tZ0gQUTkglY5XksBDNboWBQK3gZQQfhwDwuurpM319D2/ri7V1rPtKxlnZGYmQl07NMSjx64Ey3O7sfChSomeu89L2YkejO0qmVw7yhPS285g0QU3rTK0geQACDfk305q1FHfJ1CtqveAAAgAElEQVRNp7eZhXrl7Dpt2SLlVVepWYCurqEnldcd3SuNZuCdOSPlsmVSDh4s62YkYot8GvdfOCPRl4rsjmZU2vJ3pmdhobrGvXu7f1+6a4tOgLMauXEz/Ba0WY1CiGlCiFwhRG5xcXGwTkvu+No7wl4Czzi7TosXA99/r5LLXV3DggIgKanuZ7O5rjCpq3vlqGfHwyE76xqJ48YBrVoBkyYBO3eqdRG3fV6MvKy/44HYl9Aeh/3rTXM0o9KWvz13c+YA332nroG73jJ3bSEi0kjQZjVKKZcAWAKo5PpgnZfc8CeA0tPMQj2xT6S3vU69eqkhRasdO9RXi0XtY3sNHSWgWxPP3d0rRzPwZs50OmvPOiMxJwd4+21VY7RFC+B3v1OxT91yPXHAWwEKuO2vW3y8CjLLyoCoKO+qzNtfN/sZjEREoaJV1xk41GhcBhlmMQxXi0R7M0Rr3ddkkueH2C65RMr27b27V06G7CxR0fLHH6W87z4p27ZVD19yiWrOhx9KWVnp5HiBer84um6JiaphY8Z4fq7CQrWfdWwUUN+PGmX4oXBwqJEbN8Nv2h2IgRfpka9Vyn3haU6Sp/lXhYVSxsdLKYR/uVp2wV5BdIqc232VTE2ulIAqMD90qJQ5OVKePev94f3mSR6bt/ldM2Zc/NwQVJrXGgMvbtyMv2lVTmIFgG8BdBZCHBFCTNXiuER+02JGnKflDDzNSfKkAnpRkRqWLCoCUlP9q5YeH48TDdrgpfLbkSm+xeXnduDRvFvRvFWDC2Ykjh/vxRqJWnJ03UaNUjW2fM3vOn5c1US79Va1JSay0jwR6YImOV5SyvFaHIcCrD4VO3VXpdwb7iqaW3maL+euArp927dvV0WyvGx7aamqFZqTA6z/5CnUSDO6JZfj6WbvYlyTj9Dh48UeHyugHF23Vq1UP5WvuWRMjicineJajfWJXoudalUg05YWtax8qVXm73p+riq6e9B2pzMSHzSrNRJ3x+CBb0e4D7oCcU9ccXTd/JiZSUSkW6EY32SOV5A5y6ERQh/Jxq6S0f3hby2rUNQqs57TNjEckHLyZKdPqamR8ssvpZw2TcqmTdXuLVqoXPSvv5bSYvGhHd7ck2Dm0QXqvWIQYI4XN26G39jjVR846v2x1oYKZe9XoKvf+9v75G2pDS16Y6znrKlRpRSEANLSgDNnLthN1q6ReP/9QIcO6rTLlwM33gh88AFQWKhGMfv2xYULU7vjyz0JRk8qV0ogojDBwKs+sA0gAFUXae9e9ekdyg+wAC1tc54WRTG9Cd60CkCOH1c1tzZvBrKzgeTk823fvx944gkgPV2lfVnXSMzJqVsj8cYbXayR6I6zRPcePS5+7cEMhgL9XgkWDpUSUSi62TjUGALW2kvr10uZlFQ3lBXqpX40WtomqOyH1vxd2saN48elfPFFKfv0qTt0v35SLlwoZXGxJqe4kP09SUtzfG+CPRRrxPeKPT+HSsGhRm7cDL+xx6u+sPb+DBwIXHed+vzWw1I//g4HBkpREdC7N9Cnz8Vtsu/ZCkBvTGkp8MYbwJAhQOvWwB/+APz6K/D008CBA6rK/IwZqrq85qz3xGJR2/btjnu0gr1slF7fK57gUCkRWYUi2mOPV4ixUr172dnyosKbrnq27HtjJk3yOuG8okLKtWulHDtWypgYdegOHaR86CEpt20LzMt0yZMeLXfvpWAm3uuZRr2DYI8XN26G30JyUgZe5FSoP6hdVVGPinL+4WkfgCQmejSk5GhGYvPmdTMSa2qC9Lqd8TegrOezEC+gwVApAy9u3Iy/caiR9CXUtcYKClTFdLO57jGzWSWYHzjgfGjNOpTbu7cqsrp//8VDSrWJ1XJLHvJ63Y4H7jqLhASVa/3mm2pY8YMP1G7WGYmmUP8LtR/e++orz+6Pq6G1+ppgbuShUiLSTiiiPfZ4GVCge6ICnKDuFVfr/HkytOakV6xgwqNyrnhEpkbtVWskmqrlTTeFcI1Eb3h7f1wNrbEXzGdgjxc3bobfQv3/afJFKHoMAt0TpadyAa7W+XNXosIu4fxEeSPMz2mKzPgCXL58Dh6Vc9GsoggLMQNFllZ4/wOB8bfHqDUS9dwT5O39cZR4v3KlepwJ5kRUjzHwMqJgDscFazZWsGfIubJ6tQo0Vq1SW0GBVzXASo+U4M3fvoIhvY6jNY7iLvkizl7aFk9HPIoD6ICvcDVmYDFaxJRdGLyEepjVFUf3x2wGxo51HijaD61df71+gmsiohBh4GUkgQqCXPW0BLMnysA5MJWVwHvvAePHA60+y8HEz6Zgx+HGuO8BM7ZuBbaOewoPWJ5CB/NR9QSzGaioUMFMYmLoSg1408vmbb6Xfe/ghx/qJ7gmIgoRBl5GEqggyFlPS1GRWm05IiIwH5b2H/paVJp3dNwAsViADRtUDBIfDwwfDqxfD0yZoi7n/v3AU08BXbuiLmgZMEAtATRgQF1wGcphVm962TyZQOCOgYNrIiJNhCKxjMn1ftCyere7hGlrEnRiovZ1vwoLpYyPVwt1a51kbZ+8reHEAItFXYb775eyXTt5Pm88K0vK99+XsrLSxwMHuyq7s3sfFeX+WoVi8XCSUkom13PjFgabJj1eQojBQojdQoh9QogHtTgmOaFlj4GznhYpLxz62r9f9XD07u1fT5RVTIwqx15U5Hq9SG97rpwNxbZr53fu1P79wJNP1q2R+NxzQLduamHqEyfU15tu8mONxGD3BDm79+PGub9WesrHIyIyGn8jNwBmAD8BuBxAJIA8AKmunmOoHq9QF/QMJNteJ9uelkD2aDjraTGZLj6+t2UH7NvtqhCqB06ckPKll6TMzKx7ar9+Ui5YEKA1EoPNtpfN2bVyVi6Cqx+EBNjjxY2b4TcteryuBLBPSlkgpawEsBLAzRocVx/0PNPMX3PmqF6l1NQLe1oC2aPx7bdAXNyFBUoBYOLEuuP7OonAvt1CAElJdb061nOOG+f0EGfPqmKmN96oDnfXXWrdxKeeqlsjMTs7QGskBpttL9vkyUDbtp7nmWmVj0dEVM9oEXi1AXDY5ucjtY8ZWzgvamv72gC1CHKPHsCrr9Z9gAZq6GvJEqC4GKipUYGQECrZ/MyZun38STa3bXd2NlBVBZSVqd/V1Kivr79+wb20nZHYsqWKAbdvB+67D2pG4lbgwQeBDh20uQS6YRs8LV0KDB3K4UMiogDTIvASDh6TF+0kxDQhRK4QIre4uFiD0waYngp6as2T16Z1j4Z9sAeoQMhkApKTLzy+Pz1u9u3u2VP15gwZomZn1r5eS9Zt2JBzxOGMxK++spuRGG6c5c5pEWzruQgsEZEOaBF4HQHQzubntgAK7XeSUi6RUmZIKTPi4uI0OG2A+frhb4QPHm9em1avx1mwd+SI46BOqx631atVb06HDpA1FuRFXoEHyv6ChLX/hwGjmuONN4DBg4H331cvdcECoF8/HayRGEjOhs/9Cbat75OHHgrfoXkiIi34myQGIAJAAYBE1CXXp7l6jmGS631JIDbKOnSevjYtX0+wSybU2r9fyidS35RpzY7WrZHY6nu5fLkB1kjUUiDXwzSbA3dsOg9MrufGzfCbkPKiUUGvCSFuBPA81AzHV6WUT7jaPyMjQ+bm5vp9Xl2JiVE9R/aio4Hy8uC3x1+BeD2jRqnetmnTVK5XUVHAkrKLi4G33gJycoCNG9Vj/fqpTrcxY+yS462FYletCu+cpqIiYNYs4N13Vd5bbCwwciTw7LO+v25n7xNA9Wj6c2y6iBBis5QyI9TtICLfaTKgIqX8UEqZLKXs6C7o8oueh/HCLScsEK8nwDPhHM1IPHNG5Wrt3+9iRmI4z1y1FYjZqvbvE6Bu0kSokvP1/HeCiOo9Y2Wy6PkDMtyKShrk9bibkbhtm5qRmJDg4Mm+zFw1+oe61rNVbd8n1sS4UaNUhBuqa6TnvxNEVO9pMtToLa+HGo0yjBfEobSg0OnrsViAb75R1eLffhs4dQpo3hy49VbV+ZKZ6WFyvC9DbzNnAosXA9Onq0x8Ct37xH6I2Je/EwYbZuZQI1EYCEVimdfJ9XpfGy6cq9sHi5traLFImZcXwjUSA5mYTr5xtCant38njDIZphaYXM+Nm+E3Yww16n3Yi0Mb/nNyDQ8cqKun1b37hWskHj/uwxqJ9kOFng69hVsOn5E5GyK+/HLP/06Ec4FkItI1YwReQPAXEfYE/3j7z8E1LBZxWNDgT+jbF0hMBB5+GLg0thILOj2Hwh+P4/33VQzUsKEP57MP8DxN+Nd78F+fFBSopaCsbINgBtJEpHPGyPHSq0BMz69vaq/h2XfWY035ICw3TcI6ORA10oz0dPVZOG4ckPB3P3OrtMgT1GnOW73i7D6azUB1tXfHys5W9zEyUs3SMEDeHnO8iIzPOD1eesReEL9UVgLvb47H+P/djZblB3AblmO7pQvu6/Fp3YzEv8UgIVGDXkVPezhczVrkwtChZ72P1tkT0dGq9+v6670/lh570Yko7Bkz8NLTlH7+8XbN7l5ZLHX1tOLjgWHDgPWHOmNKWi6+em0f9mc/g6cSFtetkajVkJCnQbLe8/X09N4PBet9BNR9rKwEBg4EPvzQ+2MxkCaiEIgIdQN8YvvhGOqhAds/1vPnB++8RpkGX3uvtv7pFeQkPoIVK4BDh1T8dPPNKoYaNKgxIiOvVvtPeenC52vZq2gNkm2HCq3sh7AWLlSb3kqW6Om9Hyqu7iMRkc4ZK8dLL/W89BD06L2eVEwMDpxrhRUYjxxkIR9dYUY1bjCtx4Q3hmD4cC+S4z3NrfLnvug9X08v730KKeZ4ERmfsYYa9TITKZTDUTqfSVlcrOLAvuklSMQBPIyn0AQlmN/gbhSN/gM+ONrT+xmJng4J+XNf9J6vp5f3PhER+cVYgVeoPxz1EPTo8AP47Nm6elrx8cDvfw+UlEfiySvfxX5xOb6OHoSZNS8irqUIzL3S6r7oOV8v1O99IiLShLECLyC0H456CHqC8QHsQQJ3ZSXO19Nq1Qq47TYgP1+N1m3dqr5/qM0yJGQPCfy9KihQw4LWmW6+3he9J1vrOTAkIiKPGC+5PlTJ7IB+eh28SS72Je/JSQK3dY3EnBy1RuLJk0CzZsCkSSrOuWiNxGDdq/h4YPdu1UCzOXx7g0L53iciIk0YK7leD4xWRNObJHwnCdxbIzOQc8/3F81IzMpS5ZMiIwPUdk9ERwMVFRc/bjIBNTXBbw9RADG5nsj4GHiFK19mwdnM7DtQFocVDSYjJ/YO5Je0g9kM3HCDCrZuvtnH5XoCYfJkYNkyICJCVS4P1GxEPcxk9ZSR2kpeYeBFZHx+5XgJIcYIIbYLISxCCP4x0BMf8tGKI+Kx4NBQ9Cv7RM1IrPobmsRUYP589Vn+wQfqELoIuqwJ9cuWqZ+ty8WUlQVmmFHvhVVtGamtRET1jL/J9fkARgHYoEFb6jetK5J7mI9mOyOxdWvg91+Pxy/NLseTfyjC/tsew9d97sfMmUBcXIhfjz37wNJsBm68UfWAaXlOPcxk9ZSR2kpEVE/5FXhJKXdKKXdr1Zh6LRC9FI5mwRUVoar/tXj/jdMXzEjctg3485+BvDwg/2RrPPRCPBLemON5/pp9oBXoXhf7wFJKoEMHYOlSbXPu3PUc6mkJHz3MuiUiIpeMN6sx3ARyqRqbAMTy4nw1I3HgBry94y2c/Lrp+RmJWVlA3752MxK9ZQ202ra9MKnd+nqEAAoLtR0CDMbSMe56DvW0hI9eZt0SEZFTbpPrhRCfAnD0l/sRKeWa2n2+ADBLSuk0Y14IMQ3ANABo3759r4MHD/ra5vCi5VI1DpKqt25V5R9WzDuEQ2iPGJRhBN5FFnJwPdYhMtrsX4DnLInfKjYWaNMG2LdPBUmhDk584Wgm60cf6XMJH6PNuiWvMLmeKAxIKf3eAHwBIMPT/Xv16iXJxowZUppMUkZHq6/Z2Y73KyyU8uqrpSwqcvz77GwpTSa5f8Ij8sknpUxPlxKQ0myW8sbryuWbfV6SpTFx6sHYWCknTHB+LE8VFkqZlaWOZz1uUpKUQqifHW3R0f6d09N2ubpWWhzf/nX7cj0D3U4KKwBypQZ/s7lx4xa6zXiV6z2lp9wbdzytSO4sbyomBj+LFliwEOhn+RKJy+fi4YeBJts31s1I/DQaE7rno2HFSdfDUN5eN0fDW9XVQHY2sH49kJSkEt8B73KO/L1/wc4x83VYjzMQiYjqF3+iNgAjARwBUAHgOIBPPHleUHq8ant/nPYeGUl0tHTUc1Qa1VwuX656syJElQSkTMM2+WSD2bLg5rvrelGsvSpDhkg5c6aUW7aoryNHXnwuX67byJHOj+tpb54W7ZDS6bUKSC+bq9etp3ZS2AB7vLhxM/wWkpMGNPAKxw80m2GtSkTI96NGyvEdvpGxMTUSkLJdOykf6PGxzBPdpSXKQYDjSRDjy3XzZJjM2+DE3/un1RBgoBmlnaQrDLy4cTP+Fn5DjWE4pd7SKh5f/9oT2WX/QDyKMLRiNT451g0TJ5mwYQNw4ADwdOJidMvuC/GdzXClN3WdfLlungyTebvwtL/3zygz+4zSTiIi0lT4lZMIow+0bdtUcVO1RuIsxERUYsTAX5FlWojrG/wXkYveqtvZ0QLKrmZM2vPmugWyBIYW98+TMhN6WFYnGOUwiIhIV8Iv8AIM/YF28KAKtJYvB/LzVV769dcDTz4J3HxzJBo2jASQXbu54W0Q42nAkpKiVssuK1OBlquAzhf+3j9HQag9PdTf8qSdREQUVrhItg78/DPw9tsq2PrmG/VY5hWVmDAlEmPGuFmux13PjdZ1nWbOVL1bVtHRQGUlMH26MWp0+bJ4uD/00LNGYYN1vIiML/xyvILNx7IHZ/cdQ07KHAwddA7x8RIzZwKni6vwxJVrUCA64puMuz1bI9FdnpW3OVbO2OaL2Tp3Tj1uhLIdQPBzAFkugoiIbDDw8pezD1YHAVlVFfDBB+pzv1VKU0zY9Ri2fn4S93b5EHnohvw9kXj4fyOQKAvcL3Ac7AWRCwrUcKK1Jhegvh81CjhyxDjV0YOVA8gFq4mIyAEGXq646s1y98FaG5BZ/jYHX3+tRuji44GhQ4FPVpzExOrXsAH9ccDSDvPyh6IbtkHYHt9dT0ywe27i49WK2rbrMNbUqMeMNoTmacFaf4Th7FoiIvJfeCbXa8VVAnZBwcUzBgcPBt55BxAC25COHMzFikXjcXAREIMy3DwuFhOGnML1U9ogUlY4P29UlPuemFDM3jx+HEhMBK64Qv38/ffGGWK0FYyk9jCaXUtERNoxfo+Xq14p6+/y8rzLw/JkmMjBB+vBbWfwtHwA3SJ3ohu24Rnch1TTbryRuRAn9pVixQpg6KRmiLzt1ovPaR3CS0sDvvvOs56YYPTc2Fq9WgWcq1apraAgOEOMRlr+yVaw7w8REeme8Wc1zpwJLF7seFad9XcpKcCOHaq34Ycf3Pc6uKp/ZfvcUaPw86Wd8PaycuTU3Iqv0R8AkIlvkIUc3Iq3EGc6Bdx2m6pyap3ZNmoUsGePapP1+lunL/o76zAcubrHRPUIZzUSGZ9xAy9XZQEAx7+z3cdd6YDsbFV+ITLyonIJv/4KrFkD5OQAn3yi1oRObXIEE8r+hfFVy5BoOgR07Kh6ylavBj76SBXosg0ctC7zEI6CXfqBSOcYeBEZn3GHGh0lL48aBfTooYZ2srKczyA7d8797DK7YaKqwmJ88IHKj27ZUn3duhW4915gyxYgf9wTeLhmLhKja4eTBg5UmfQLFqiEavshS63KPOiJ1kOCTFAnIqIwY9zAyzbHKipKDQlu3w78739qWKpxY6CiAjDZvcSICM8+vFevhuXF+fi6tDtmyvmI//ptDB0KfPwxMHEizq+ROG+eip3ECQf5PL4GDloEMKHIi9K6ZhUT1ImIKMwYN/AC6nqlqqrUz7t31/UsLVqkepeuuQZo2lT93mxW44JuPrzz/1uMh9q/ics71KB/f2DpUmDQIGDtWhXPLFoE9O9UBNNvbQIbRz1YvgYOWgQwwSzcGciaVUxQJyKiMGLcHC/AeQ4QoHqWrMnwHuRTWddIzMlRi1ObUY3r2+9C1hPpGDECaNjQ7vieJnx7k8ulRU5TKPKiPJ2MQER+YY4XURiQUgZ969Wrl9REYaGUWVlSxsZKqeYHSmk2SymElNnZbp9eXCzlggVS9utX9/RMfC1fwkx5HHF1D0ZH1z0pOrrucdvNdh+tXk9srJQTJkhZVBTcY/hixgwpTSZ1HUwmj64/EXkHQK4Mwd9sbty4abf5NdQohHhGCLFLCLFVCPGOEOJSjeJBz9gO5VlzuUaNUjMSnQxJ/fqr6tUaOlQ9feZM4NQp4IkngIJNJ/BN1gL8PnYpWqLYcU5WIBO+tchpio9XQ6plZZ4VYtUKhwSJiIjc8rdy/XoAD0kpq4UQ8wA8BOAB/5vlBesHvu1Qnl018qoqYP16YPnyutGwdu3UjMSsLKBbN5WSBLR0H/gEOuHb0evx1tdfq6/Dh9fVBgu0YFSDB9RrGTeuriYaERGRgWiW4yWEGAngFinlBHf7alpA1QmLBfj2WxVsvfUWcPIk0KyZqlOalQX063fxhEcAnuVkhaIGlycBR32oe8ViqlSPMceLyPi0DLzeA7BKSvmmu30DGXjl56tga8UKlTAfEwPcfLMKtm64QdVDNSRPAo5wTnKvD0ElkRsMvIiMz22OlxDiUyFEvoPtZpt9HgFQDWC5i+NME0LkCiFyi4uLtWl9rYMHgaefVkOGXbsCzzyjVgl64w01crdiBTBsmJ9BV6jWC/SmVEM4171iMVUiIgoDbgMvKeVAKWW6g20NAAghJgMYCmCCdNF9JqVcIqXMkFJmxMXF+d3wn39WMUj//kBCAvDQQ0CjRsBLLwGFhWqVnttuU495xVmAFcy6WLa8DTj0kOQeiCA1nINKIiKqN/xKrhdCDIZKph8gpSzTpknO/fqrKmK6fLnNGompakbi+PFAYqIGJ7ENsBYsuHiIa+FCtQVriMvbgCNYSe6u2F9DrWgx8YCIiCiE/MrxEkLsAxAF4GTtQ5uklDPcPc/XHK+JE4E33wTatlWdQBfOSPSSfbK6sxyiqChg9OjQ5k0ZZUFt5mERBRRzvIiMz686XlLKTlLKdlLKHrWb26DLH3/+M/Dllyqn6/waib4EXcDFQ4fOhvQOHAjdEJd1yG7BAmMsqM08LCIiIpcMtVZjjx7A1Vc7KQPhKWfJ6pdf7jzAClXe1Jw5wFdfAb/5jTEKkjIPi4iIyCVDBV6acNUr4yzAsi6A3bKlqlcR6PpRtsGhlKrnKz5em0WnA00Pyf1EREQ65W/leuNx1SvjLjE9UEnj9goKVCKbxXLh4+fOqeBLz/lSekjuJyIi0qn61+MFeN8r400tLVc8LbMQH6964WxFRLjOlwpVnTEiIiLyWPgFXp4EINahQ9tkdVfP0ypp3JtaYGfPAmlpKsAzm1XtDFf5UqGqM0ZEREQeC7/Ay9cAxNXz/E0a96XHbPVqIDkZyM4GNm9WSwY5Cgq16o0jIiKigNNsrUZvBGStRl9rSHn6PH9qaQVyDcVwXp+RiC7AOl5Exhc+PV6+Dgd6+jxHw5OeCmSZBZZwICIiMozwCbx8DUCCFbgEsswCSzj4jpMSiIgoiMJnqBHwfTjQKEvykPZmzgQWLwamTw98fTYiP3Gokcj4wivwIvIU15UkA2LgRWR84TPUSOQNritJREQhwMCL6idOSiAiohBg4EX1FyclEBFRkNW/tRqJrLiuJBERBRl7vIKN5QuIiIjqLb8CLyHEHCHEViHEFiHEOiFEa60aFra4piIREVG95W+P1zNSym5Syh4A3gcwW4M2BU8we5+4piIREVG951fgJaU8Y/PjJQCCXxTMH8HsfWL5AiIionrP7+R6IcQTACYBKAHwW79bFAz2xTMXLlRbIItnsnwBERFRvee2x0sI8akQIt/BdjMASCkfkVK2A7AcwF0ujjNNCJErhMgtLi7W7hX4IlS9TyxfQEREVK+57fGSUg708Fg5AD4A8Bcnx1kCYAmglgzytIEBEareJ5YvICIiqtf8ndWYZPPjcAC7/GtOELH3iYiIiILM3xyvp4UQnQFYABwEMMP/JgVJsHqfioqAceOAVauYz8VrQURE9Zy/sxpHSynTa0tKDJNSHtWqYWGDdbvq8FoQEVE9J6QMfrpVRkaGzM3NDfp5g8p+5qRVIGdO6hWvBZEmhBCbpZQZoW4HEfmOSwYFCut21eG1ICIiAsDAK3BYt6sOrwUREREABl6BxZmTdXgtiIiImONFRGQUzPEiMj72eBEREREFCQMvIiIioiAxbuBVVAQMGMBcISIiIjIM4wZeLMZJREREBmO8wCsmBhACWLgQsFjUVyHU40REREQ6ZrzAy99inByiJCIiohAxXuDlbzFODlESERFRiBgv8AJ8K8bJIUoiIiIKsYhQN8Anq1fXfT9/vmfPKSgAZs0C3n0XKCtTQ5QjRwLPPhuYNhIRERHZMWaPly+4XiARERGFWP0JvACuF0hEREQhpclQoxBiFoBnAMRJKX/W4pgB4csQJREREZFG/O7xEkK0AzAIwCH/m0NEREQUvrQYavwngPsBSA2ORURERBS2/Aq8hBDDARyVUuZp1B4iIiKisOU2x0sI8SkAR1P/HgHwMIDrPTmREGIagGkA0L59ey+aSERERBQehJS+jRAKIboC+AxAWe1DbQEUArhSSulyumBGRobMzc316bxERPWVEGKzlDIj1O0gIt/5PKtRSrkNQEvrz0KIAwAydD2rkYiIiCiE6lcdLyIiIqIQ0izwklImGLa3q6gIGDCABVWJiIgooNjjBQBz5gBffw08/nioW0JERERhrH4HXjExgBDAwoWAxaK+CqEeJ8nyziYAAAYVSURBVCIiItJY/Q68CgqArCwgNlb9HBsLTJgA7N8f2nYRERFRWKrfgVd8PNC4MXDuHBAdrb42bgxc5qhsGREREZF/6nfgBQDHjwMzZgCbNqmvTLAnIiKiAPG5jlfYWL267vv580PXDiIiIgp77PEiIiIiChIGXkRERERBwsCLiIiIKEgYeBEREREFCQMvIiIioiBh4EVEREQUJEJKGfyTClEM4KCPT28BQI+LcbNd3mG7vMN2eUev7QL8a1sHKWWclo0houAKSeDlDyFErpQyI9TtsMd2eYft8g7b5R29tgvQd9uIKPA41EhEREQUJAy8iIiIiILEiIHXklA3wAm2yztsl3fYLu/otV2AvttGRAFmuBwvIiIiIqMyYo8XERERkSHpPvASQjwjhNglhNgqhHhHCHGpk/0GCyF2CyH2CSEeDEK7xgghtgshLEIIpzOUhBAHhBDbhBBbhBC5OmpXsK9XMyHEeiHE3tqvTZ3sV1N7rbYIIdYGsD0uX78QIkoIsar2998JIRIC1RYv2zVFCFFsc43uCFK7XhVCnBBC5Dv5vRBCvFDb7q1CiN/opF3XCCFKbK7X7CC0qZ0Q4r9CiJ21/xb/5GCfkFwvItIBKaWuNwDXA4io/X4egHkO9jED+AnA5QAiAeQBSA1wu1IAdAbwBYAMF/sdANAiiNfLbbtCdL3+DuDB2u8fdHQfa393NgjXyO3rBzATwKLa78cBWKWTdk0B8FKw3k82570awG8A5Dv5/Y0APgIgAPQG8J1O2nUNgPeDfK3iAfym9vtGAPY4uI8huV7cuHEL/ab7Hi8p5TopZXXtj5sAtHWw25UA9kkpC6SUlQBWArg5wO3aKaXcHchz+MLDdgX9etUe//Xa718HMCLA53PFk9dv295/A7hOCCF00K6QkFJuAHDKxS43A1gmlU0ALhVCxOugXUEnpSySUv5Q+30pgJ0A2tjtFpLrRUShp/vAy87tUP9LtNcGwGGbn4/g4j90oSIBrBNCbBZCTAt1Y2qF4nq1klIWAeqDCUBLJ/tFCyFyhRCbhBCBCs48ef3n96kN/EsANA9Qe7xpFwCMrh2e+rcQol2A2+QpPf8b7COEyBNCfCSESAvmiWuHqHsC+M7uV3q+XkQUQBGhbgAACCE+BXCZg189IqVcU7vPIwCqASx3dAgHj/k9XdOTdnmgr5SyUAjREsB6IcSu2v+lh7JdQb9eXhymfe31uhzA50KIbVLKn/xtmx1PXn9ArpEbnpzzPQArpJQVQogZUL1y1wa4XZ4IxfXyxA9Qy+ycFULcCOBdAEnBOLEQoiGA/wC4W0p5xv7XDp6ih+tFRAGmi8BLSjnQ1e+FEJMBDAVwnZTS0R+nIwBs/+ffFkBhoNvl4TEKa7+eEEK8AzWc5FfgpUG7gn69hBDHhRDxUsqi2iGVE06OYb1eBUKIL6B6C7QOvDx5/dZ9jgghIgA0QeCHtNy2S0p50ubHf0HlPepBQN5T/rINeKSUHwohFgghWkgpA7qOoxCiAVTQtVxKudrBLrq8XkQUeLofahRCDAbwAIDhUsoyJ7t9DyBJCJEohIiESoYO2Iw4TwkhLhFCNLJ+DzVRwOHsqyALxfVaC2By7feTAVzUMyeEaCqEiKr9vgWAvgB2BKAtnrx+2/beAuBzJ0F/UNtllwc0HCp/SA/WAphUO1uvN4AS69ByKAkhLrPm5gkhroT6m3fS9bP8PqcA8AqAnVLK55zspsvrRURBEOrsfncbgH1QuRBbajfrTLPWAD602e9GqNlDP0ENuQW6XSOh/tdaAeA4gE/s2wU1Oy2vdtuul3aF6Ho1B/AZgL21X5vVPp4B4OXa7zMBbKu9XtsATA1gey56/QAehwrwASAawNu177//Abg80NfIw3Y9VfteygPwXwBdgtSuFQCKAFTVvr+mApgBYEbt7wWA+bXt3gYXM32D3K67bK7XJgCZQWhTP6hhw602f7du1MP14saNW+g3Vq4nIiIiChLdDzUSERERhQsGXkRERERBwsCLiIiIKEgYeBEREREFCQMvIiIioiBh4EVEREQUJAy8iIiIiIKEgRcRERFRkPw/laEpt6lVrn8AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Algoritmo\n",
    "\n",
    "# Matriz de 1s para servir como as previsões\n",
    "ones = np.ones(num_points)\n",
    "\n",
    "# Convertendo as variáveis para matrizes NumPy\n",
    "X_matrix = np.matrix(np.column_stack((X_entrada, ones)))  # Matriz aumentada (explicada nas aulas)\n",
    "y_matrix = np.transpose(np.matrix(y_saida))\n",
    "\n",
    "# Print shapes\n",
    "print(np.shape(X_matrix))\n",
    "print(np.shape(y_matrix))\n",
    "\n",
    "# Encontrando a transposta das matrizes e aplicando o dot product\n",
    "XtX = X_matrix.transpose().dot(X_matrix)\n",
    "Xty = X_matrix.transpose().dot(y_matrix)\n",
    "\n",
    "# Print shapes\n",
    "print(np.shape(XtX))\n",
    "print(np.shape(Xty))\n",
    "\n",
    "# Inversa\n",
    "solution = np.linalg.inv(XtX).dot(Xty).tolist()\n",
    "print(\"\\nCoeficientes da Regressão:\")\n",
    "print(solution)\n",
    "\n",
    "# Coeficientes\n",
    "w, b = solution\n",
    "\n",
    "# Previsões\n",
    "y_pred = w[0]*X_entrada + b[0]\n",
    "print(\"\\n\", y_pred)\n",
    "\n",
    "# Plot\n",
    "plt.plot(X_entrada, y_saida, 'r*', label='Dados')\n",
    "plt.plot(X_entrada, y_pred, 'b', label='Linha de Regressão - Previsões NumPy')\n",
    "plt.legend(loc=\"upper left\", bbox_to_anchor=(1,1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Exercício 10 (Super Desafio) - Mesmo exercício anterior, mas agora com o TensorFlow.\n",
    "\n",
    "Dica: Deixamos algumas linhas de código para ajudá-lo nesta tarefa."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "ERROR:root:Internal Python error in the inspect module.\n",
      "Below is the traceback from this internal error.\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3326, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-2-09bdfa551d48>\", line 8, in <module>\n",
      "    X_entrada = np.linspace(-2, 2, num_points)\n",
      "NameError: name 'np' is not defined\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2040, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'NameError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 18, in swig_import_helper\n",
      "    fp, pathname, description = imp.find_module('_pywrap_tensorflow_internal', [dirname(__file__)])\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/imp.py\", line 296, in find_module\n",
      "    raise ImportError(_ERR_MSG.format(name), name=name)\n",
      "ImportError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\", line 58, in <module>\n",
      "    from tensorflow.python.pywrap_tensorflow_internal import *\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 28, in <module>\n",
      "    _pywrap_tensorflow_internal = swig_import_helper()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 20, in swig_import_helper\n",
      "    import _pywrap_tensorflow_internal\n",
      "ModuleNotFoundError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 1101, in get_records\n",
      "    return _fixed_getinnerframes(etb, number_of_lines_of_context, tb_offset)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 319, in wrapped\n",
      "    return f(*args, **kwargs)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/ultratb.py\", line 353, in _fixed_getinnerframes\n",
      "    records = fix_frame_records_filenames(inspect.getinnerframes(etb, context))\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/inspect.py\", line 1502, in getinnerframes\n",
      "    frameinfo = (tb.tb_frame,) + getframeinfo(tb, context)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/inspect.py\", line 1460, in getframeinfo\n",
      "    filename = getsourcefile(frame) or getfile(frame)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/inspect.py\", line 696, in getsourcefile\n",
      "    if getattr(getmodule(object, filename), '__loader__', None) is not None:\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/inspect.py\", line 733, in getmodule\n",
      "    if ismodule(module) and hasattr(module, '__file__'):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/importlib/__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 953, in _find_and_load_unlocked\n",
      "  File \"<frozen importlib._bootstrap>\", line 219, in _call_with_frames_removed\n",
      "  File \"<frozen importlib._bootstrap>\", line 1006, in _gcd_import\n",
      "  File \"<frozen importlib._bootstrap>\", line 983, in _find_and_load\n",
      "  File \"<frozen importlib._bootstrap>\", line 967, in _find_and_load_unlocked\n",
      "  File \"<frozen importlib._bootstrap>\", line 677, in _load_unlocked\n",
      "  File \"<frozen importlib._bootstrap_external>\", line 728, in exec_module\n",
      "  File \"<frozen importlib._bootstrap>\", line 219, in _call_with_frames_removed\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/__init__.py\", line 42, in <module>\n",
      "    from . _api.v2 import audio\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/_api/v2/audio/__init__.py\", line 10, in <module>\n",
      "    from tensorflow.python.ops.gen_audio_ops import decode_wav\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/ops/gen_audio_ops.py\", line 10, in <module>\n",
      "    from tensorflow.python import pywrap_tensorflow as _pywrap_tensorflow\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 50, in __getattr__\n",
      "    module = self._load()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow/__init__.py\", line 44, in _load\n",
      "    module = _importlib.import_module(self.__name__)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/importlib/__init__.py\", line 127, in import_module\n",
      "    return _bootstrap._gcd_import(name[level:], package, level)\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/__init__.py\", line 49, in <module>\n",
      "    from tensorflow.python import pywrap_tensorflow\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\", line 74, in <module>\n",
      "    raise ImportError(msg)\n",
      "ImportError: Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 3326, in run_code\n",
      "    exec(code_obj, self.user_global_ns, self.user_ns)\n",
      "  File \"<ipython-input-2-09bdfa551d48>\", line 8, in <module>\n",
      "    X_entrada = np.linspace(-2, 2, num_points)\n",
      "NameError: name 'np' is not defined\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/IPython/core/interactiveshell.py\", line 2040, in showtraceback\n",
      "    stb = value._render_traceback_()\n",
      "AttributeError: 'NameError' object has no attribute '_render_traceback_'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 18, in swig_import_helper\n",
      "    fp, pathname, description = imp.find_module('_pywrap_tensorflow_internal', [dirname(__file__)])\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/imp.py\", line 296, in find_module\n",
      "    raise ImportError(_ERR_MSG.format(name), name=name)\n",
      "ImportError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "During handling of the above exception, another exception occurred:\n",
      "\n",
      "Traceback (most recent call last):\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow.py\", line 58, in <module>\n",
      "    from tensorflow.python.pywrap_tensorflow_internal import *\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 28, in <module>\n",
      "    _pywrap_tensorflow_internal = swig_import_helper()\n",
      "  File \"/Users/marcelonunesalves/opt/anaconda3/lib/python3.7/site-packages/tensorflow_core/python/pywrap_tensorflow_internal.py\", line 20, in swig_import_helper\n",
      "    import _pywrap_tensorflow_internal\n",
      "ModuleNotFoundError: No module named '_pywrap_tensorflow_internal'\n",
      "\n",
      "\n",
      "Failed to load the native TensorFlow runtime.\n",
      "\n",
      "See https://www.tensorflow.org/install/errors\n",
      "\n",
      "for some common reasons and solutions.  Include the entire stack trace\n",
      "above this error message when asking for help.\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'np' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m"
     ]
    }
   ],
   "source": [
    "# Solução\n",
    "\n",
    "# Import\n",
    "import tensorflow as tf\n",
    "\n",
    "# Variáveis x e y\n",
    "num_points = 200\n",
    "X_entrada = np.linspace(-2, 2, num_points)\n",
    "y_saida = X_entrada + np.random.normal(0, 1, num_points)\n",
    "\n",
    "# Matrizes (tensores com o TF)\n",
    "X_tensor = tf.constant(X_matrix)\n",
    "y_tensor = tf.constant(y_matrix)\n",
    "\n",
    "# Transposta\n",
    "XtX_tensor = tf.matmul(tf.transpose(X_tensor), X_tensor)\n",
    "Xty_tensor = tf.matmul(tf.transpose(X_tensor), y_tensor)\n",
    "\n",
    "# Solução\n",
    "solution_tf = tf.matmul(tf.matrix_inverse(XtX_tensor), Xty_tensor)\n",
    "\n",
    "# Abre a sessão TF\n",
    "sess = tf.Session()\n",
    "soln_tf = sess.run(solution_tf).tolist()\n",
    "\n",
    "# Encontra os coeficeintes\n",
    "w_tf, b_tf = soln_tf\n",
    "\n",
    "# Previsões\n",
    "y_pred_tf = w_tf[0]*X_entrada + b_tf[0]\n",
    "\n",
    "# Plot\n",
    "plt.plot(X_entrada, y_saida, 'r*', label='Dados')\n",
    "plt.plot(X_entrada, y_pred, 'b', label='Linha de Regressão - Previsões NumPy')\n",
    "plt.plot(X_entrada, y_pred_tf, 'g', label='Linha de Regressão - Previsões TF')\n",
    "plt.legend(loc=\"upper left\", bbox_to_anchor=(1,1))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fim"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
